[WIP] No-overlay mode enhancement (#5289)
ovn-kubernetes/ovn-kubernetes: docs/okeps/okep-5259-no-overlay.md

  encapsulation/decapsulation.

## Terminology

* **Primary Network** - The network that is used as the default gateway for the
  pod. Typically recognized as the eth0 interface in the pod.
* **Cluster Default Network** - This is the routed OVN network that pods attach
  to by default today as their primary network. The pod default route, service
  access, as well as kubelet probe are all served by the interface (typically
  eth0) on this network.
* **User-Defined Network** - A network that may be primary or secondary, but is
  declared by the user.
* **ClusterUserDefinedNetwork** - Also known as CUDN, this is a cluster-scoped
  User-Defined Network created by the cluster admin that can be consumed by
  different namespaces.
* **Layer 2 Type Network** - An OVN-Kubernetes topology rendered into OVN where
  pods all connect to the same distributed logical switch (layer 2 segment)
  that spans all nodes.
* **Layer 3 Type Network** - An OVN-Kubernetes topology rendered into OVN where
  pods have a per-node logical switch and subnet. Routing is used for pod to pod
  communication across nodes. This is the network type used by the cluster
  default network today.

> * **Localnet Type Network** - An OVN-Kubernetes topology rendered into OVN where

Author: I copied this line from https://github.com/pliurh/ovn-kubernetes/blob/882f4451fc0e30973b03a7f1009b80d7b5e854b6/docs/okeps/okep-5193-user-defined-networks.md#L31-L32.
And according to https://github.com/ovn-kubernetes/ovn-kubernetes/blob/d56ed1edf7a9705cf8e579b40d50cc2bfdec560b/go-controller/pkg/ovn/base_secondary_layer2_network_controller.go#L181, there is no transit switch for a Localnet network.

  pods connect to a per-node logical switch that is directly wired to the
  underlay.
* **BGP-routed Egress Traffic** - The pod egress traffic that is routed via BGP
  routes.
* **Non-BGP-routed Egress Traffic** - The pod egress traffic that is routed via
  non-BGP routes (static, direct, etc.).

## Goals

* Support no-overlay mode for the cluster default network(CDN).

> * Support no-overlay mode for Primary layer-3 ClusterUserDefinedNetworks

Reviewer: OK; then please add that to the enhancement.

  pod-to-external traffic, OVN-Kubernetes shall adjust its SNAT behavior so that
  pod IPs are exposed for traffic routed via BGP, but SNAT is still applied for
  non-BGP-routed egress traffic.

* For east-west traffic:
  * Intra-node traffic remains unchanged.
  * Inter-node traffic will follow the same path as north-south traffic.
    Pod-to-pod, Pod-to-clusterIP traffic will be forwarded by the provider
    network without encapsulation and NAT. Pod-to-node egress traffic will be
    SNATed to the node IP at the source node. This is to allow a pod to access
    nodePort services in a different UDN.

For the default network, we will add a new flag
`--default-network-encapsulation` to the `ovnkube` binary. This flag accepts
either `none` or `geneve`, with `geneve` being the default. Setting
`--default-network-encapsulation=none` will configure the default network to
operate in no-overlay mode. If the `--default-network-encapsulation=none` flag
is set but the required RouteAdvertisements or FRRConfiguration CR is missing,
`ovnkube` will exit with an error.

For CUDNs, no-overlay mode shall be configured via the CUDN CRs.

> ### API Details

Author: Here's the assumption: 
1. If the egress traffic is routed via an external BGP route, it implies that the pod subnet is learnt by the external network. So we don't need SNAT to the node IP.
2. If the egress traffic is routed via a non-BGP route, it implies that the pod subnet is unknow to the external network. So we shall SNAT it to the node IP.

> ### API Details

Reviewer: Right - I do understand that. 

What I wonder is, if the assumption will match the expectations of the entire user base. Hence, I would suggest to do nothing, and offload this to a separate enhancement.

Point being, I'd rather rely on user feedback than on assumptions - which are often disastrous. 

Just a thought.

> ### API Details

Author: Added a new field to RA to make it configurable.

    EncapsulationProtocolGeneve EncapsulationProtocol = "Geneve"
)

// EncapsulationConfiguration defines the encapsulation configuration of a Layer3 network
type EncapsulationConfiguration struct {
    // +kubebuilder:validation:Enum=None;Geneve
    // +kubebuilder:default=Geneve
    Protocol EncapsulationProtocol `json:"protocol"`
}

type Layer3Config struct {
    ...
    // +optional
    EncapConfig EncapsulationConfiguration `json:"encapsulationConfiguration,omitempty"`
}
```

The `spec` field of a ClusterUserDefinedNetwork CR is immutable. Therefore, the
encapsulation configuration cannot be changed after the
ClusterUserDefinedNetwork CR is created.

The `Layer3Config` is shared between CUDN and UDN. Once we add the new
`encapsulationConfiguration` field, it will be available for both CUDN and UDN.
To prevent a user from setting this field in a UDN CR accidentally, we will add
the following CEL validation rule to the UserDefinedNetwork CRD.

```yaml
    layer3:
      description: Layer3 is the Layer3 topology configuration.
      x-kubernetes-validations:
      - message: Setting layer3.encapsulationConfiguration is not supported for UserDefinedNetworks

>         rule: '!has(self.encapsulationConfiguration)'

Reviewer 2: the rule translates into: "this UDN is valid if it doesn't have `encapsulationConfiguration` set"

>         rule: '!has(self.encapsulationConfiguration)'

Reviewer: Maybe I wasn't clear enough, sorry for that. I fully understand what this CEL expression does. What I do not understand is how this CEL expression can be used to achieve the goal you have set, which, IIUC, is to allow `encapsulationConfiguration` to only be available on C-UDNs.

Using your own words: since the `Layer3Config` is shared between CUDN and UDN, once you add the new `encapsulationConfiguration` field how can this CEL expression - which is mapped to `Layer3Config` - prevent adding an `encapsulationConfiguration` to a UDN, but allow it for a C-UDN ?

Surely I'm missing something.

>         rule: '!has(self.encapsulationConfiguration)'

Author: This CEL is only added to UDN. So it means in a UDN CR is valid if it doesn't have encapsulationConfiguration set. It prevents user from setting `encapsulationConfiguration` for a UDN.

>         rule: '!has(self.encapsulationConfiguration)'

Author: This CEL will only be added to the UDN CRD. So the CUDN will not be affected.

>         rule: '!has(self.encapsulationConfiguration)'

Reviewer: Ah, OK. Does that the `Layer3Config` will no longer be shared between CUDN and UDN - each will have different attributes ? If so, IMHO, your text should be updated. 

If not, would you please explain how can you define CEL expressions on a shared type that will be added to only one of the CRDs ?

>         rule: '!has(self.encapsulationConfiguration)'

Reviewer: Right, thank you for literally pointing it out.

need to add the static node subnet routes into the ovn_cluster_router. The
traffic shall be forwarded following the source-based route instead. The nexthop
depends on the gateway mode of the cluster.

#### Traffic Path in Shared Gateway Mode

At the ovn_cluster_router, the pod-to-remote-pod egress traffic will be
forwarded to the GatewayRouter via the join switch. Unlike the current route
advertisement behavior, the routes to remote node subnets will be imported to
the gateway router. Once the traffic reaches the GatewayRouter, it will be
forwarded to the nexthop node according to the imported BGP routes.

Here's an example routing table on a UDN's ovn_cluster_router, the pod subnet of
the node is 22.100.1.0/24. The LRSR routes all the pod egress traffic to the
gateway router according to source-based route:

```
IPv4 Routes
Route Table <main>:
                10.96.0.1                22.100.1.2 dst-ip
               10.96.0.10                22.100.1.2 dst-ip
               100.65.0.3                100.65.0.3 dst-ip
            22.100.1.0/24                100.65.0.3 src-ip
            22.100.0.0/16                100.65.0.3 src-ip

> ```

Author: This is the routing table of ovn_cluster_router which doesn't contains any BGP routes. I put it here to show that thte pod-to-remote-pod traffic will not be routed to the transit swith in the no-overlay mode.

    ovn_cluster_router->>Pod1 on Node A: Packet to local Pod1 (10.128.0.5) via the node switch
```

#### Import Routes to NodeSubnets

Changes to BGP behavior are necessary to import node subnet routes into the
gateway router and the host routing table of each node. A new `encapsulation`
field will be introduced to the NAD spec that indicates the encapsulation of the
network. If a network is operating in no-overlay mode, this field will be set to
`none`. We don't support users modifying this field in a NAD manually to enable
no-overlay mode. This information can then be passed into the NetInfo object and
utilized by both the RouteAdvertisement controller and the route import
controller.

An example OVN-Kubernetes NAD may look like:

```yaml
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: l3-network
  namespace: default
spec:
  config: |
    {
            "cniVersion": "0.3.1",
            "name": "l3-network",
            "type": "ovn-k8s-cni-overlay",
            "topology":"layer3",
            "mtu": 1500,
            "netAttachDefName": "default/l3-network",
            "role": "primary",

>             "encapsulation": "none"

Author: Users shall not set the encapsulation feild in a NAD, it shall be generated from a CUDN CR by OVN-K.

router, which ensures node-to-pod traffic traverses the OVN logical topology.
However, we need the routes to be imported into the gateway router in no-overlay
mode. We need to update the route import controller to check whether a network
is in no-overlay mode and import the routes from the host routing table if so.

#### Conditional SNAT for Egress Traffic

The OVN-Kubernetes BGP feature currently disables SNAT for all pod egress
traffic if the pod network is advertised. The only exception is for traffic
destined to other nodes in the cluster (pod-to-other-node), to ensure nodePort
services can be accessed across networks. This egress traffic will remain SNATed
to the node IP before leaving the node.

However, this approach assumes that all external destinations can route back to
the pod IPs, which is often not true for destinations reached via non-BGP routes
(like a default internet gateway). So, instead of disabling SNAT for all pod
egress traffic, we shall do the following:

* For egress traffic forwarded via a BGP route, SNAT shall be disabled.
* For egress traffic forwarded via a non-BGP route, the egress traffic shall be

>   SNATed to the node IP.

Reviewer 3: This seams really complicated to pull off. Is it really a hard requirement? Could we just expose a field in the RA that says "SNAT egress traffic" or something? If true, then we only add pod subnets to the address set and skip SNAT if they are in there (positive match), all other traffic is SNAT'ed.

gateway_port        : []
logical_ip          : "10.128.0.7" # A local pod IP
logical_port        : []
match               : "ip4.dst!=$skip_snat_dst||ip4.dst==$snat_dst"
options             : {stateless="false"}
priority            : 0
type                : snat
```

##### Local Gateway Mode

To maintain SNAT for non-BGP-routed egress traffic, the masquerade nftables rule
will be kept. We will introduce a new controller. It will watch the host route
updates, and insert an nftables rule for each imported BGP route to bypass the
SNAT rule.

If users add custom routes to the host routing table that overlap with BGP route
prefixes and want to retain SNAT for traffic routed through these custom routes,
they must also insert corresponding masquerade rules.

In the following example, `10.128.0.0/24` is the local nodeSubnet,
`174.16.1.0/24` is the CIDR of a BGP route.

```text
table ip nat {
    chain postrouting {
        type nat hook postrouting priority 100; policy accept;
        # skip SNAT for traffic via BGP routers
        ip saddr 10.128.0.0/24 ip daddr 174.16.1.0/24 return
        ip saddr 10.128.0.0/24 masquerade
    }
}
```

#### UDN Traffic Isolation

> Currently, for cross-UDN pod-to-pod traffic, OVN-Kubernetes supports 2 modes:

Author: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5276

  * Gateway Modes: LGW and SGW
  * IP Modes: IPv4, IPv6, and dual-stack
  * BGP Topologies: iBGP with an external route reflector, full-mesh iBGP,
    iBGP+eBGP combined and eBGP only.
  * test suite: conformance, control-plane
* API Testing Details
  * ClusterUserDefineNetworks reports expected status conditions
  * No overlay mode cannot be enabled for UserDefineNetworks
  * EgreeIP and EgressService CR reports expected status conditions
* Scale Testing Details
  * The impact of the number of the imported BGP routes
* Performance Testing Details
  * Test pod to pod throughput and latency
* Cross-Feature Testing Details - coverage for interaction with other features
  * UDN isolation between UDNs in no-overlay mode
    * Default mode
    * Loose mode
  * UDN isolation between UDNs in no-overlay and overlay mode
    * Default mode
    * Loose mode
  * NetworkPolicy
  * EgressFirewall

## Risks, Known Limitations and Mitigations

In the no-overlay mode, east-west traffic relies on the BGP network. Therefore,
internal and external BGP speaker outages may impact cluster networking.

All the limitations of the BGP integration feature also apply to the
no-overlay feature.

> As OVNI is no longer delivering pod-to-pod traffic end-to-end, it will

Reviewer 3: nit: s/OVNI/OVN

necessitate BGP knowledge for debugging.

## OVN Kubernetes Version Skew

To be discussed.

## Alternatives

N/A

> ## References

Author: For now, EVPN integration is not in the scope of this enhancement.

encapsulation configuration cannot be changed after the
ClusterUserDefinedNetwork CR is created.

The `Layer3Config` is shared between CUDN and UDN. Once we add the new
`encapsulationConfiguration` field, it will be available for both CUDN and UDN.
To prevent a user from setting this field in a UDN CR accidentally, we will add
the following CEL validation rule to the UserDefinedNetwork CRD.

```yaml
    layer3:
      description: Layer3 is the Layer3 topology configuration.
      x-kubernetes-validations:
      - message: Setting layer3.encapsulationConfiguration is not supported for UserDefinedNetworks
        rule: '!has(self.encapsulationConfiguration)'
```

To give users the control of pod outbound traffic SNAT, we will add a new
`natOutbound` field to the RouteAdvertisements CRD. If it's `true`. The
pod‑to‑external traffic will be SNATed to the node IP before leaving the node.
The field defaults to `false`.

```golang
type RouteAdvertisementsSpec struct {
  ...
  // NatOutbound, when set to true, enables SNAT for pod egress traffic to non-pod subnets. Defaults to false.
  // +optional
  // +kubebuilder:default=false
  NatOutbound bool `json:"natOutbound,omitempty"`

> ```

Reviewer 2: We discussed this briefly on Monday during the BGP call and I don't have a solid opinion about this yet. But here's a few thoughts.

First of all, I think we need a wide consensus on the fact that we need `NatOutbound`. I do prefer this approach to the previous one, where the SNAT behaviour depended on the route a packet would hit on egress: the behaviour is clear from the start and can be decided by the cluster admin.

As for where this parameter should be, I think it makes sense to have all parameters for the same feature grouped together, wherever possible. Does it make sense for `NatOutbound` to be configured in the RA and the no-overlay mode to be enabled 1) as an input parameter for the default network 2) in the CUDN spec for whatever CUDN we want to affect? 
Would it make more sense to have it on a per-network basis instead, along with the field(s) that enable no-overlay? 
If the naming `encapsulation` that we chose for the spec makes it difficult to add `NatOutbound` underneath, maybe we can rename it to something else, like `OverlayMode` and model the values accordingly. 

It's just a thought. If it makes more sense to have it under RA, I'm fine with that too!

* Support no-overlay mode for Primary layer-3 ClusterUserDefinedNetworks
  (CUDNs).
* A cluster can have networks operating in overlay and no-overlay modes
  simultaneously.
* Use the BGP feature to exchange routes to node subnets across the cluster.
* Allow direct communication without any overlay encapsulation for east-west
  traffic.
* Maintain compatibility with existing OVN-Kubernetes features where applicable.
* Compatible with both local gateway and shared gateway modes.

## Future Goals

* When OVN-Kubernetes supports BGP for UserDefinedNetwork CRs, extend no-overlay
  mode support for Primary layer-3 UDNs.

* Support toggling no-overlay mode on/off for an existing network. To support
  this functionality, there are two things we need to consider.
  First, in no-overlay mode, as there is no Geneve overhead, a network can have
  the same MTU size as the provider network. However, overlay networks require a
  smaller MTU (default 1400). We cannot modify a pod interface MTU without
  restarting the pod. So OVN-Kubernetes must work with other external
  controllers to accommodate such a migration.
  Secondly, to facilitate a smooth, node-by-node migration, it's crucial to

>   allow networks to operate in a hybrid no-overlay and overlay mode.  

Author: > If users would be willing to accept temporary disruption, would this alleviate the concern?

Yes. But we'd better add a status indicator to the CUDN API to let users know when the transition is complete.

I think the options to bypass MTU migration are valid. However, until we get more precise customer requirements, I believe it would be best to leave it as a future goal.

external BGP routers.

## Proposed Solution

The core idea is to leverage the existing BGP feature to advertise the IP subnet
allocated to each node (which contains the IPs of pods running on that node) to
an internal/external BGP route reflector. The BGP route reflector will then
populate these routes throughout the physical network, allowing other nodes to
directly route traffic to pod IPs without needing an overlay. Within a cluster,
different networks can operate in different overlay modes. Users can have
overlay and no-overlay networks simultaneously.

When no-overlay mode is enabled for a network:

* For north-south traffic, we will follow the existing implementation of the BGP
  feature; the traffic from pods will no longer be SNATed on egress. In the BGP
  feature, SNAT is disabled for all pod egress traffic that leaves the node,
  regardless of whether the destination route is BGP learnt or not. For users
  who only want to have no-overlay mode but do not want to expose pod IPs
  externally, this behavior is not ideal. Instead of disabling SNAT for all
  pod-to-external traffic, we shall let users to choose whether SNAT shall be

>   skipped for the pod egress traffic by adding a new `natOutbound` field to

Reviewer 3: So this is a use case where you want to use BGP just for east/west traffic, and have no external BGP peer for N/S. Sorry I missed that in your Story 3. If we want to include it in the enhancement then that is OK, but I still think the API needs some work. We can discuss that in a separate comment though.

  to SNAT pod-to-external traffic to the node IP. Pod-to-remote-pod traffic is
  not SNATed, regardless of this setting.

For CUDNs, no-overlay mode shall be configured via the CUDN CRs.

### API Details

We introduce new fields to the `spec.network` section of the
ClusterUserDefinedNetwork (CUDN) CRD to control network encapsulation:

* **`overlay`**: Specifies the encapsulation method.
  * Supported values: `Geneve` (default) or `None` (no-overlay).
  * More overlay encapsulation methods can be supported in the future.
* **`noOverlayOptions`**: Contains configuration for no-overlay mode. This is
  required when `overlay` is `None`.
  * `outboundSNAT`: Defines the SNAT behavior for outbound traffic from pods.
    * Supported values: `Enable` or `Disable`. Defaults to `Disable`.
    * `Enable`: SNATs pod-to-external traffic to the node IP. This is useful
      when pod IPs are not routable on the external network.
    * `Disable`: Does not SNAT pod-to-external traffic. This requires pod IPs to
      be routable on the external network.
    * Pod-to-remote-pod traffic is never SNATed in no-overlay mode.

>   * `routeAdvertisements`: Specifies the name of the RouteAdvertisements CR used

Reviewer 3: >No, the RA here is only for advertising podNetwork for enabling no-overlay. It is not for configuration but more for cross-check between CUDN and RA.

I mean't a use case for enabling NAT without no overlay mode is when you only want to advertise EgressIPs and not the pod network with a BGP RA. Maybe that is already implied today with the current RA behavior. Need to check.

>   * `routeAdvertisements`: Specifies the name of the RouteAdvertisements CR used

Reviewer 4: > I meant a use case for enabling NAT without no overlay mode is when you only want to advertise EgressIPs and not the pod network with a BGP RA. Maybe that is already implied today with the current RA behavior. Need to check.

Yes, advertising just EIPs will keep the NAT in place.

For every option that we add to the RA we need to think how we are going to aggregate it if multiple RAs select the same network providing each different values of this option. NAT is something that we either do for a network or we don't do for a network. So on that scenario we need to either report an error or make an assumption: for example, as long as one of the RA request the NAT, then it stays regardless of what the other RAs are requesting. This requires specific handling on our side and it stays confusing for a user. 
Whereas a NAT option on the network itself seems more straight forward to understand I think.

>   * `routeAdvertisements`: Specifies the name of the RouteAdvertisements CR used

Reviewer 4: I expressed the idea of specifying an RA in the CUDN [here](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5289#discussion_r2345388797):

```
Both the no overlay and disable snat options assume there is something else configured that would make that work, like for example an RA. We should be able to do that, just assume there is something and not be concerned about it. However I have been thinking it could be beneficial for the user to specify what those RAs are. That could be like a three-way acknowledgement between the RA, the network and the user that the configuration that needs to be in place, is indeed in place. Additionally, we could use the references to those RAs to perform additional validation and report network scoped problems in different ways: failing to create the network, network status conditions, network events...
```

So it is basically used for validation. If we don't like the idea, we can drop it and just trust the user is doing what it needs to be doing.

>   * `routeAdvertisements`: Specifies the name of the RouteAdvertisements CR used

Reviewer 4: That's a question that wouldn't need to be answered if we had an API that promoted acknowledgement and validation :tongue: 

In any case, I understand how it can be confusing and I am ready to drop my suggestion and move on.

>   * `routeAdvertisements`: Specifies the name of the RouteAdvertisements CR used

Author: > What's the consequent state we will be in if the RA doesn't exist selecting the network but user created a no-overlay network?

We shall still allow the network to be created. The north-south traffic works, whereas the east-west routing doesn't. It allows users to create RA after CUDN or switch an RA for a network.

Currently, the advertisement controller sets the annotation `k8s.ovn.org/route-advertisements` to NADs. It shall also set the annotation to CUDNs. Then the CUDN controller can use it as reference to validate the RA configuration and check the RA status for no-overlay mode.

>   * `routeAdvertisements`: Specifies the name of the RouteAdvertisements CR used

Reviewer 5: > Currently, the advertisement controller sets the annotation `k8s.ovn.org/route-advertisements` to NADs. It shall also set the annotation to CUDNs. Then the CUDN controller can use it as reference to validate the RA configuration and check the RA status for no-overlay mode.

ah! good point - @pliurh is this mentioned in the implementation details? perhaps I missed it.

type SnatOption string

const (
    OverlayProtocolNone   OverlayProtocol = "None"
    OverlayProtocolGeneve OverlayProtocol = "Geneve"

    SnatEnable  SnatOption = "Enable"
    SnatDisable SnatOption = "Disable"
)

// NoOverlayOptions contains configuration options for networks operating in no-overlay mode.
type NoOverlayOptions struct {
    // OutboundSNAT defines the SNAT behavior for outbound traffic from pods.
    // +kubebuilder:validation:Enum=Enable;Disable
    // +kubebuilder:default=Disable
    // +optional
    OutboundSNAT SnatOption `json:"outboundSNAT,omitempty"`
    // RouteAdvertisements specifies the name of the RouteAdvertisements CR
    // used to advertise the pod network of this CUDN.
    // +optional
    RouteAdvertisements string `json:"routeAdvertisements,omitempty"`
}

type NetworkSpec struct {
    ...
    // Overlay describes the overlay protocol for east-west traffic.
    // Allowed values are "None" and "Geneve".
    // - "None": The network operates in no-overlay mode.
    // - "Geneve": The network uses Geneve overlay.
    // Defaults to "Geneve".
    // +kubebuilder:validation:Enum=None;Geneve
    // +kubebuilder:default=Geneve
    // +optional

>     Overlay OverlayProtocol `json:"overlay,omitempty"`

Author: The proposed design follows the same pattern as the `topology` config in the UDN CRDs. And it aligned with the [Kube API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#automatic-resource-allocation-and-deallocation).

To support EVPN, it would be like this:
```
    topology: Layer3
    layer3:
      role: Primary
    overlay: EVPN              # Geneve | None | EVPN
    evpnOptions:
        evpnRef:
            name: blue-evpn     # name of the EVPN CR that defines macVRF/ipVRF
```

>     Overlay OverlayProtocol `json:"overlay,omitempty"`

Reviewer 3: ok makes sense

    // It is required when Overlay is "None".
    // +optional
    NoOverlayOptions *NoOverlayOptions `json:"noOverlayOptions,omitempty"`
}
```

To prevent a user from setting unsupported no-overlay configurations in CUDN
accidentally, we will add the following CEL validation rules to the
`spec.network` field of the ClusterUserDefinedNetwork CRD.

```yaml
    network:
      x-kubernetes-validations:
      - message: "overlay 'None' is only supported for Layer3 topology"
        rule: 'self.overlay != "None" || self.topology == "Layer3"'
      - message: "noOverlayOptions is required when overlay is 'None'"
        rule: 'self.overlay != "None" || has(self.noOverlayOptions)'
      - message: "noOverlayOptions is only supported for no-overlay networks"
        rule: 'self.overlay == "None" || !has(self.noOverlayOptions)'
      - message: "noOverlayOptions.routeAdvertisements is required when overlay is 'None'"

>         rule: 'self.overlay != "None" || (has(self.noOverlayOptions) && has(self.noOverlayOptions.routeAdvertisements) && self.noOverlayOptions.routeAdvertisements != "")'

Reviewer 2: Uhm... if we make this `RouteAdvertisements` field required, what value should it have if we want to support the simplest use case we discussed, where the cluster admin only wants to enable no-overlay for all eligible networks between nodes and doesn't want to export pod subnets to the outside? In this use case we're in charge of creating the base FRRConfig and the RA.

external_mac        : []
external_port_range : "32768-60999"
gateway_port        : []
logical_ip          : "10.128.0.7"         # The local pod IP
logical_port        : []
match               : ""
options             : {stateless="false"}
priority            : 0
type                : snat
```

##### Local Gateway Mode

In the following example, `10.128.0.0/24` is the pod subnet in the local node,
`10.128.0.0/16` is the CIDR of the pod network. `natOutbound` is enabled in for
the network.

```text
table ip nat {
    chain postrouting {
        type nat hook postrouting priority 100; policy accept;
        # skip SNAT for pod-to-pod traffic
        ip saddr 10.128.0.0/24 ip daddr 10.128.0.0/16 return
        ip saddr 10.128.0.0/24 masquerade
    }
}
```

#### UDN Traffic Isolation

Currently, for cross-UDN pod-to-pod traffic, OVN-Kubernetes supports 2 modes:
loose and strict. In the loose mode, cross-UDN traffic is allowed if an external
router can route the traffic correctly. In strict mode, cross-UDN traffic is
blocked by ACL rule defined in the node switch. In no-overlay mode, these two
isolation modes shall still operate in the same way.

### Workflow

> #### Enable No-overlay Mode for the Default Network

Reviewer 2: So this is the complete workflow for external reflectors. Did we drop the idea of internal route reflectors that we evoked two weeks ago in order to avoid a full mesh? Or are we leaving this to the downstream usage of the feature? We also talked about having the AS number configurable, in a similar way that calico does.

   instances by OVN-Kubernetes.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:
      name: default
      namespace: frr-k8s-system
      labels:
        # A custom label
        network: default
    spec:
      bgp:
        routers:
        - asn: 64512
          neighbors:
          # the external BGP route reflector
          - address: 172.20.0.2
            asn: 64512
            disableMP: true
    ```

3. The cluster admin shall create a RouteAdvertisements CR to advertise the
   default network. Make sure the RouteAdvertisements CR selects the
   FRRConfiguration defined in the previous step.

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: RouteAdvertisements
    metadata:
      name: default
    spec:
      advertisements:
      - PodNetwork
      # Select the FRRConfiguration defined in step-2 with the custom label.
      frrConfigurationSelector:
        matchLabels:
          network: default
      networkSelectors:
      - networkSelectionType: DefaultNetwork
      # The empty nodeSelector selects all nodes. We don't support a network in a overlay and no-overlay hybrid mode.

>       nodeSelector: {}

Reviewer 2: Do we report an error somewhere if the nodeSelector isn't empty?

Author: ovnkube-cluster-manager shall print error to its log.

* The `--default-network-overlay` flag accepts either `none` or `geneve`,
  defaulting to `geneve`. Setting it to `none` configures the default network to
  operate in no-overlay mode. More overlay encapsulation methods can be
  supported in the future. If `--default-network-overlay=none` is set but no
  `RouteAdvertisements` CR is configured for the default network, `ovnkube` will
  exit with an error.
* The `--default-network-nat-outbound` flag can only be set when
  `--default-network-overlay=none`. It accepts `Enable` or `Disable` and
  defaults to `Disable`. When set to `Enable`, it configures the default network
  to SNAT pod-to-external traffic to the node IP. Pod-to-remote-pod traffic is not
  SNATed, regardless of this setting.

For CUDNs, no-overlay mode is configured via their respective CRs. Due to a BGP
limitation, CUDNs with overlapping pod subnets cannot be advertised to the same
VRF. To enable no-overlay mode for these networks, a VRF-lite topology is required.

### API Details

We introduce new fields to the `spec.network` section of the
ClusterUserDefinedNetwork (CUDN) CRD to control network encapsulation:

* **`overlay`**: Specifies the encapsulation method.

>   * Supported values: `Geneve` (default) or `None` (no-overlay).

Reviewer 5: @pliurh : thinking out loud.. on calling it `no-overlay` -> today we assume its no-overlay with BGP since scope of RA is just for BGP - although nothing prevents us from adding OSPF tomorrow and the name `RouteAvertisement` is generic enough that I don't know if we might support more protocols. Or maybe it would be new API `LinkStateAdvertisement`
In that case, if we keep no-overlay as the name, then inside noOverlayOptions I hope we can have enough granularity for each type of no-overlay? OR do we want to be more specific around this?

external_ip         : "172.18.0.2"         # Node/GR external IP used for SNAT
external_mac        : []
external_port_range : "32768-60999"
gateway_port        : []
logical_ip          : "10.128.0.7"         # The local pod IP
logical_port        : []
match               : ""
options             : {stateless="false"}
priority            : 0
type                : snat
```

##### Local Gateway Mode

In the following example, `10.128.0.0/24` is the pod subnet in the local node,
`10.128.0.0/16` is the CIDR of the pod network. `outboundSNAT` is enabled in for
the network.

```text
table ip nat {
    chain postrouting {
        type nat hook postrouting priority 100; policy accept;
        # skip SNAT for pod-to-pod traffic
        ip saddr 10.128.0.0/24 ip daddr 10.128.0.0/16 return
        ip saddr 10.128.0.0/24 masquerade
    }
}
```

#### UDN Traffic Isolation

Currently, for cross-UDN pod-to-pod traffic, OVN-Kubernetes supports 2 modes:
loose and strict. In the loose mode, cross-UDN traffic is allowed if an external
router can route the traffic correctly. In strict mode, cross-UDN traffic is
blocked by ACL rule defined in the node switch. In no-overlay mode, these two

> isolation modes shall still operate in the same way.

Reviewer 5: I think we agreed to make this a future goal thing right? So I can resolve this I think? and then https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5289/files#r2577174668
comment is same as this - there was some workaround @trozet had mentioned users can do at router config level for this.

   But it relies on the K8S-FRR project to support the route reflector feature in
   the future.

2. ovnkube-cluster-manager will create a base FRRConfiguration CR named
   `default` in the `frr-k8s-system` namespace, and a RouteAdvertisements CR
   named `default` that advertises the default network. The generated
   FRRConfiguration will configure BGP speakers on each node to advertise the
   pod subnets and set up BGP sessions between nodes.

   When nodes are added or removed, ovnkube-cluster-manager shall update the base
   FRRConfiguration CR accordingly, and the route advertisement controller will
   regenerate the per-node FRRConfiguration instances.

   For example, if the cluster has 3 nodes with pod subnets
   `10.128.0.0/24`, `10.128.1.0/24`, and `10.128.2.0/24`. The generated
   RouteAdvertisements CR will look like:

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: RouteAdvertisements
    metadata:
      name: default
    spec:
      advertisements:
      - PodNetwork
      frrConfigurationSelector:
        matchLabels:
          network: default
      networkSelectors:
      - networkSelectionType: DefaultNetwork
      nodeSelector: {}
    ```

> 

Reviewer 4: Yeah, but that's an important change because then we would have to generate FRRConfigurations per node with a node selector.

    kind: FRRConfiguration
    metadata:
      name: default
      namespace: frr-k8s-system
    spec:
      bgp:
        routers:
        - asn: 64514
          neighbors:
          # BGP peers with other nodes
          - address: <node1-ip>
            asn: 64514
            toReceive:
              # Allow to receive the routes to remote pod subnets.
              # This is no-overlay specific behavior.
              allowed:
                prefixes:
                - ge: 24
                  le: 24
                  prefix: 10.128.0.0/16
          - address: <node2-ip>
            asn: 64514
            toReceive:
              allowed:
                prefixes:
                - ge: 24
                  le: 24
                  prefix: 10.128.0.0/16
          - address: <node3-ip>
            asn: 64514
            toReceive:
              allowed:
                prefixes:
                - ge: 24
                  le: 24
                  prefix: 10.128.0.0/16
    ```

  The following steps are the same as steps 5 and 6 in the previous section.

#### Create a ClusterUserDefinedNetwork in No-overlay Mode with Unmanaged Routing

Here is a configuration example:

> 1. The frr-k8s pods shall be deployed on each node.

Author: It truly doesn't matter. Let's leave it as it is.

ClusterUserDefinedNetwork CRD. When enabled, pod egress traffic will be SNATed
to the node IP before leaving the node.

The `outboundSNAT` field is intended for deployments where a user wants a
no-overlay network but has not advertised the pod network to be routable from
external networks. When enabled, pod-to-external traffic is SNATed and forwarded
by a default gateway, while pod-to-remote-pod traffic is not SNATed regardless
of the value of this field.

##### Shared Gateway Mode

If `outboundSNAT` is enabled for a network, in the gateway router, the default
SNAT rule will be kept for pod-to-external traffic. But pod-to-remote-pod
traffic remains unaffected by SNAT. Therefore, the network controllers in
ovnkube-controller will need to add an address set for the pod subnets of each
CUDN and the default network. Then the network controllers will add an exemption
rule with the address set to the SNAT entries. We use the `exempted_ext_ips`
field instead of adding a negative match to the `match` field, as using a
negative match in an OVN SNAT rule is not efficient. Setting the
`exempted_ext_ips` field results in one OpenFlow rule being created for each pod

> SNAT entry in the OVS bridge.

Reviewer 4: nit:

The phrase starts with `If `outboundSNAT` is enabled for a network` and that's important. If the network has geneve transport or `outboundSNAT` disabled, we shouldn't be doing any of this to avoid the risk of impacting scale. I am not sure if this could be conveyed more clearly. Same for local gateway mode where this is even less explicitly stated.

   cluster internal BGP fabric. The cluster admin only needs to start
   ovn-kubernetes with the following flags:

    ```ini
    [bgp]
    topology = full-mesh
    as-number = 64514
    ```

   The generated FRRConfiguration will be the same as the one described in the
   managed default network section.

3. A cluster admin wants to enable no-overlay mode for the blue network by
   creating the following ClusterUserDefinedNetwork CR.

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: ClusterUserDefinedNetwork
    metadata:
      name: blue
      labels:
        network: blue
    spec:
      namespaceSelector:
        matchExpressions:
        - key: kubernetes.io/metadata.name
          operator: In
          values: ["ns1", "ns2"]
      network:
        topology: Layer3
        layer3:
          role: Primary
          # The UDN MTU shall not be larger than the provider network's MTU.
          mtu: 1500
          subnets:
          - cidr: 10.10.0.0/16
            hostSubnet: 24
        transport: "NoOverlay"
        noOverlayOptions:
          outboundSNAT: "Enable"
          routing: "Managed"
    ```

4. The UDN controller in ovnkube-cluster-manager will create a

>    RouteAdvertisements CR named `blue` that advertises the blue network.

Reviewer 4: Again, maybe important to avoid generating common RouteAdvertisements CR names to avoid the potential for conflict.

      direction LR
      A(Node A)
      B(Node B)
      C(Node C)
      A <-- iBGP --> B
      B <-- iBGP --> C
      C <-- iBGP --> A
    end
```

Here's an example FRRConfiguration:

```yaml
apiVersion: frrk8s.metallb.io/v1beta1
kind: FRRConfiguration
metadata:
  labels:
    network: default
  name: receive-all
  namespace: frr-k8s-system
spec:
  bgp:
    routers:
    - asn: 64512
      # All nodes are added as neighbors to setup a full-mesh BGP topology
      neighbors:
      - address: 192.168.111.20
        asn: 64512
        disableMP: true
        toReceive:
          allowed:
            mode: all
      - address: 192.168.111.21
        asn: 64512
        disableMP: true
        toReceive:
          allowed:
            mode: all
      - address: 192.168.111.22
        asn: 64512
        disableMP: true
        toReceive:
          allowed:
            mode: all
```

##### External Route Reflectors for Larger Clusters

In a large cluster, a full-mesh BGP setup leads to more CPU and memory
consumption on the nodes. Instead of every node peering with every other node,
nodes peer only with external BGP route reflectors. This significantly reduces

> the number of BGP sessions each node needs to maintain, improving scalability.

Reviewer: what does this have to do with no overlay ? I get why this is useful, and why ovn-k should want to implement it.
I just don't see what's the relationship of that with no overlay mode. 

@pliurh

  simultaneously.
* Use the BGP feature to exchange routes to pod subnets across the cluster.
* Allow direct communication without any overlay encapsulation for east-west
  traffic.
* Maintain compatibility with existing OVN-Kubernetes features where applicable.
* Compatible with both local gateway and shared gateway modes.
* Do not compromise the existing advertised UDN isolation strict mode.

## Future Goals

* When OVN-Kubernetes supports BGP for UserDefinedNetwork CRs, extend no-overlay
  mode support for Primary layer-3 UDNs.
* Support toggling no-overlay mode on/off for an existing network. To support
  this functionality, there are two things we need to consider.
  First, in no-overlay mode, as there is no Geneve overhead, a network can have
  the same MTU size as the provider network. However, overlay networks require a
  smaller MTU (default 1400). We cannot modify a pod interface MTU without
  restarting the pod, so OVN-Kubernetes must work with other external
  controllers to accommodate such a migration.
  Secondly, to facilitate a smooth, node-by-node migration, it's crucial to
  allow networks to operate in a hybrid no-overlay and overlay mode.

> * Address the advertised UDN isolation loose mode issue when no-overlay mode is

Reviewer 4: I think @tssurya expressed interest in scoping this in

  enabled.

> ## Non-Goals

Reviewer 4: Not sure why we removed the non-goal of (not) supporting (tenant) UDNs

Author: It contradicts with future goal `When OVN-Kubernetes supports BGP for UserDefinedNetwork CRs, extend no-overlay mode support for Primary layer-3 UDNs.`

When no-overlay mode is enabled for a network:

* By default, when BGP advertisement is enabled, the BGP feature disables SNAT
  for pod egress traffic leaving nodes. For users who only want to have
  no-overlay mode but do not want to expose pod IPs externally, this behavior is
  not ideal. To address this, users will be able to enable SNAT for
  pod-to-external traffic. This will be configured via a new `outboundSNAT`
  field in the `ClusterUserDefinedNetwork` CRD for CUDNs, and the
  `no-overlay-outbound-snat` flag for the default network.

* For east-west traffic:
  * Intra-node traffic remains unchanged, handled by the local OVN bridge.
  * Inter-node traffic will follow the same path as north-south traffic, routed
    via the underlay. Pod-to-pod, Pod-to-clusterIP traffic within the same
    network will be forwarded by the provider network without encapsulation or
    NAT. Pod-to-node egress traffic will be SNATed to the node IP at the source
    node to allow a pod to access nodePort services in a different UDN.

### Configuration File and Flags

For the default network transport, the configuration file will have a new section

> `[default-network]`, with the following new flags:

Reviewer 4: There is already a section called `[default]` which contains, among other things, the MTU for the default network. Isn't that existing section appropriate to hold these parameters?

* The `transport` flag accepts either `no-overlay` or
  `geneve`, defaulting to `geneve`. Setting it to `no-overlay` configures the
  default network to operate in no-overlay mode. More overlay transport
  methods can be supported in the future. If
  `transport=no-overlay` is set, but
  no `RouteAdvertisements` CR is configured for advertising the default network,
  ovnkube-cluster-manager will emit an event and log an error message.

> * The `no-overlay-outbound-snat` flag is a string flag to configure the SNAT

Author: I'll create the `[no-overlay]` section for these flags.

  behavior for outbound traffic from pods on the default network. Supported

>   values are `enable` or `disable`. It is only used when `transport=no-overlay`.

Reviewer 4: Required? Or what is the default?

    configuration.

    One day if other routing protocols are integrated with OVN-Kubernetes, users
    can use this option to manage the routing themselves.

For no-overlay networks (the default network and CUDNs) configured with managed
routing, the configuration file will have a new section
`[internal-bgp-fabric-options]`, with the following new flags:

* The `as-number` flag is used when no-overlay networks are configured with
  managed routing. It is shared by both the cluster default network and CUDNs.
  It specifies the AS number to be used by the BGP speakers on each node for its
  default VRF. The default AS number is `64512` if not specified.

* The `topology` flag is required when no-overlay networks are configured with
  managed routing. It specifies the BGP topology to be used. Supported values
  are `full-mesh` or `route-reflector`.
  
  If the bgpTopology is set to `route-reflector`, OVN-Kubernetes will configure
  nodes with a well-known label `k8s.ovn.org/internal-bgp-role=route-reflector`
  as the route reflector, and generate the FRRConfigurations for other nodes as
  clients. This topology is more scalable for large clusters. We will work in

>   the K8S-FRR project to support the route reflector feature.

Reviewer 4: > `route-reflector` is in the scope of the enhancement. However, it will be supported once we change K8S-FRR. So `full-mesh` will be the first topology we implement.

If it is in scope, then we need to describe what is the plan to implement it FRR-K8s? I don't see the point otherwise. It is difficult to understand what's your idea of how things should behave if a user sets `route-reflector`? It wont work silently? Do you plan to return an error while it is not implemented? I don't understand why we are going down the more complicated path here instead of scoping it out.

    Node B->>Underlay Physical Network: Packet leaves Node B
    Underlay Physical Network->>Host VRF/Kernel Node A: Route packet in via breth0
    Note over Host VRF/Kernel Node A: route lookup
    Host VRF/Kernel Node A->>Host VRF/Kernel Node A: Match static route: 10.128.0.0/24 via 100.64.0.1 (Node A's pod subnet)
    Host VRF/Kernel Node A ->>ovn_cluster_router Node A: forwarded via mp0
    Note over ovn_cluster_router Node A: route lookup
    ovn_cluster_router Node A->>ovn_cluster_router Node A: Match the direct route to local pods
    ovn_cluster_router Node A->>Pod1 on Node A: Packet to local Pod1 (10.128.0.5) via the node switch
```

#### Import Routes to NodeSubnets

Changes to BGP behavior are necessary to import node subnet routes into the
gateway router and the host routing table of each node. A new `transport` key
will be included in the OVN CNI NetConf (the NAD's spec.config JSON) to indicate
the east-west transport protocol of the network. If a network operates in
no‑overlay mode, this key is set to `noOverlay`. This information can then be
passed into the NetInfo object and utilized by both the RouteAdvertisement

> controller and the route import controller of ovnkube-controller.

Reviewer 4: There is no RouteAdvertisement controller in ovnkube-controller, I think it is ok if you just mention the route import controller here

CUDN and the default network. Then the network controllers will add an
exemption rule with the address set to the SNAT entries.

In the following example, `10.128.0.0/24` is the pod subnet in the local node,
`10.128.0.0/16` is the CIDR of the pod network. `outboundSNAT` is enabled.

```text
# Address_Set for the no-overlay pod network CIDR
_uuid               : a718588e-0a9b-480e-9adb-2738e524b82d
addresses           : ["10.128.0.0/16"]
external_ids        : {ip-family=v4}
name                : a15397791517117706455

# Address_Set for the node IPs
_uuid               : e7906e7c-6aaa-46bf-8d26-e25cd650f59f
addresses           : ["172.18.0.2", "172.18.0.3", "172.18.0.4"]
external_ids        : {ip-family=v4, "k8s.ovn.org/id"="default-network-controller:EgressIP:node-ips:v4:default", "k8s.ovn.org/name"=node-ips, "k8s.ovn.org/owner-controller"=default-network-controller, "k8s.ovn.org/owner-type"=EgressIP, network=default}
name                : a712973235162149816

# per-pod SNAT rule in the Gateway Router (exclude pod-network destinations)
_uuid               : cd9264de-f2c6-4528-8f53-c2ccb5b23c8d
allowed_ext_ips     : []

> exempted_ext_ips    : ["a15397791517117706455"] # Exclude traffic to the pod Network

Reviewer 4: I would call out we are using this to avoid the negative match that would need to happen if we used `match` instead. That's the reason right?

   RouteAdvertisements CRs. It shall ensure at least one RouteAdvertisements CR
   is correctly configured to advertise the pod networks and will validate the
   RouteAdvertisements CR status. If the RouteAdvertisements CR is missing or
   its status is not accepted, it shall report event and log an error message.

7. The BGP feature is responsible for exchanging BGP routes to remote pod
   subnets between nodes. Ultimately, these routes will be imported into the
   gateway router.

#### Enable No-overlay Mode for the Default Network with Managed Routing

1. The frr-k8s pods shall be deployed on each node.

2. In the Managed routing mode, the default network controller in
   ovnkube-cluster-manager will create the RouteAdvertisements CR and the base
   FRRConfiguration CR automatically. The cluster admin only needs to start
   ovn-kubernetes with the following flags:

    ```ini
    [default-network]
    transport = no-overlay
    no-overlay-routing = managed
    no-overlay-outbound-snat = disable

    [internal-bgp-fabric-options]
    topology = full-mesh
    as-number = 64514
    ```

3. The default network controller in ovnkube-cluster-manager will create a base

>    FRRConfiguration CR named `default` in the `frr-k8s-system` namespace, and a

Reviewer 4: No, the FRRConfiguration is created by whatever controller is in charge of creating the internal BGP fabric. ItWhy would it be the default network controller in ovnkube-cluster-manager?
Also I would use more specific names than just `default` as they can collide with other instances. For example the RA controller is using `ovnk-generated-blabla` to avoid collisions

For networks configured with no-overlay mode, the OVN topology requires minor
adjustments. In the current interconnect architecture, a network can span all
cluster nodes thanks to the cross-node logical links that OVN establishes
between the transit switch and the ovn-cluster-router instances on each node.
That is the logical representation of the OVN network, and in practice, these
links are implemented through Geneve tunnels between every node in the cluster.

To route pod traffic directly over the underlay, the network's logical topology
is modified: the transit switch will not be created anymore, thus the
`ZoneInterconnectHandler` in ovnkube-controller will not create a connection to
the transit switch. As a result, static routes to remote pod subnets will not be
added to the ovn_cluster_router. Instead, traffic will be forwarded using a
source-based route, with the next hop determined by the cluster's gateway mode.
ovn_cluster_router will route all east-west traffic to the provider network,
following the same path as north-south traffic.

The following diagram shows the OVN topology of a cluster with three networks:

> * a CUDN (in green) that uses Geneve encapsulation (default behavior)

Author: Yes, that's a limitation of route advertisement, not the no-overlay mode. But you can use VRF-lite to isolate UDNs with VLANs.

  ovnkube-cluster-manager will emit an event and log an error message.

For the default network in no-overlay mode, the new flags `outbound-snat` and
`routing` will be added to the `[no-overlay]` section of the configuration file:

* The `outbound-snat` flag is a string flag to configure the SNAT
  behavior for outbound traffic from pods on the default network. Supported
  values are `enable` or `disable`. It is required when `transport=no-overlay`.

  * `enable`: SNATs pod-to-external traffic to the node IP. This is useful when
    pod IPs are not routable on the external network.
  * `disable`: Does not SNAT pod-to-external traffic. This requires pod IPs to
    be routable on the external network.

  In no-overlay mode, pod-to-remote-pod traffic is never SNATed, regardless of
  this flag's setting. However, traffic to remote nodes, the Kubernetes API
  server, and DNS is always SNATed.

* The `routing` flag is a string flag to configure
  whether the pod network routing configuration is managed by OVN-Kubernetes or
  users. Supported values are `managed` or `unmanaged`. It is required when
  `transport=no-overlay`.

>   * `managed`: OVN-Kubernetes manages the routing for the no-overlay network.

Reviewer 3: when managed mode is used, if on day 2 I decide I want to connect my workers to external BGP peers, can I do that with managed mode still enabled? Or do I need to switch to unmanaged? What happens if I switch? Does it delete all the frr-k8s configs or does it just stop managing them and leave them as is.

Can we update this okep with this information please?

    OVN-Kubernetes will generate the BGP configurations (RouteAdvertisements and
    FRRConfiguration) to advertise the pod subnets, and set up BGP sessions
    between nodes. In the managed routing mode, pod subnets will be advertised

>     to the default VRF of each node for both the default network and CUDNs.

Reviewer 3: so this "managed" config flag enables managed for all the CUDNs as well? so it is not configurable in the CUDN?

Author: This flag is for the default network only. CUDN gets its own flag in the CRD. This is a general statement for the managed mode. I should move this line somewhere else.

    Please note that the managed routing mode won't advertise pod subnets to the
    external network. Therefore, the managed routing mode is not suitable for
    clusters that have nodes in different subnets, as pod subnets are not
    routable between nodes in different subnets. So normally, this option shall

>     be used with `outbound-snat=enable`.

Reviewer 3: should this be enforced that you outbound-snat must be enable when using managed mode in the CLI and in the CUDN?

>     be used with `outbound-snat=enable`.

Reviewer 3: MEG not supported

    pod subnets. It allows users to have more control over the BGP
    configuration. Normally, this option shall be used with
    `outbound-snat=disable`.

    One day if other routing protocols are integrated with OVN-Kubernetes, users
    can use this option to manage the routing themselves.

* The `topology` flag specifies the BGP topology to use for the cluster internal
  BGP fabric, with supported values of `full-mesh` initially, more topology
  options can be added in the future. It's optional.

  This flag is required when any no-overlay network is configured with `managed`
  routing. When set, the `ovnkube-cluster-manager` generates FRRConfigurations
  to configure the BGP topology for the cluster.

  If set to `full-mesh`, ovnkube-cluster-manager will generate FRRConfigurations
  for all nodes to peer with each other in a full mesh. This topology is simpler
  to configure but less scalable for large clusters.

To configure the internal BGP fabric that CDN and CUDNs in managed no-overlay
mode will use, the configuration file will have a new section `[bgp.managed]`, with the
following new flags:

> * The `as-number` flag is used when no-overlay networks are configured with

Author: The managed mode is a per-network setting for the no-overlay mode. The user may also want to advertise another network to external routers via the default VRF.

Another use case: the user can configure the network in managed mode (handling east/west traffic by the internal BGP fabric), and additionally create manual route advertisements for the same network to handle north/south traffic via external routers.

And yes, all such configurations can be archived using the unmanaged mode alone. However, I don't think we should prevent users from doing that in managed mode.

![No-overlay topology](../images/no_overlay_topology.jpg)

#### Traffic Path in Shared Gateway Mode

At the ovn_cluster_router, the pod-to-remote-pod egress traffic will be
forwarded to the GatewayRouter via the join switch. Unlike the current route
advertisement behavior, routes to remote pod subnets will be imported to
the gateway router. Once the traffic reaches the GatewayRouter, it will be
forwarded to the nexthop node according to the imported BGP routes.

An example traffic path is as follows (assuming pod IP of 10.128.0.10 on node A,
destined to remote pod 10.128.2.11 on node B):

As the east-west traffic will not go through the transit switch, the static
routes to remote pod subnets will not be added to the ovn_cluster_router
anymore. The ovn_cluster_router routes all local pod egress traffic to the
gateway router according to a source-based route (10.128.0.0/24). The routing
table of the default network's ovn_cluster_router for node A would look like:

```text
IPv4 Routes
Route Table <main>:
               100.64.0.4                100.64.0.4 dst-ip
            10.128.0.0/24                100.64.0.4 src-ip
            10.128.0.0/16                100.64.0.4 src-ip

> ```

Author: This is an existing route. It can work with no-overlay mode without a code change.

              allowed:
                mode: filtered
                prefixes:
                # accept pod subnets of the default network explicitly
                - prefix: 10.128.0.0/16
                  ge: 24
    ```

3. The cluster admin shall create a RouteAdvertisements CR to advertise the
   default network. Make sure the RouteAdvertisements CR selects the
   FRRConfiguration defined in the previous step.

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: RouteAdvertisements
    metadata:
      name: default
    spec:
      advertisements:
      - PodNetwork
      # Select the FRRConfiguration defined in step-2 with the custom label.
      frrConfigurationSelector:
        matchLabels:
          network: default
      networkSelectors:
      - networkSelectionType: DefaultNetwork
      # The empty nodeSelector selects all nodes. We don't support a network in an overlay and no-overlay hybrid mode.
      nodeSelector: {}
    ```

4. The cluster admin enables no-overlay mode by running ovn-kubernetes with
   config file below:

    ```ini
    [default]
    transport = no-overlay

    [no-overlay]
    routing = unmanaged
    outbound-snat = disable
    ```

> 5. The default network controller in ovnkube-controller will create the OVN

Author: I think it over. We cannot skip allocating `k8s.ovn.org/node-transit-switch-port-ifaddr` because it's shared by all networks. In a cluster, the default network can be in no-overlay mode, while a CUDN can be in overlay mode.

7. The BGP feature is responsible for exchanging BGP routes to remote pod
   subnets between nodes. Ultimately, these routes will be imported into the
   gateway router.

#### Enable No-overlay Mode for the Default Network with Managed Routing

1. The frr-k8s pods shall be deployed on each node.

2. In the Managed routing mode, the default network controller in
   ovnkube-cluster-manager will create the RouteAdvertisements CR. A new
   FRRConfiguration controller will be responsible for creating the base
   FRRConfiguration CR which sets up the cluster internal BGP fabric. The
   cluster admin only needs to start ovn-kubernetes with the following flags:

    ```ini
    [default]
    transport = no-overlay

    [no-overlay]
    routing = managed
    outbound-snat = enable

    [bgp.managed]
    as-number = 64514
    topology = full-mesh
    ```

3. The default network controller in ovnkube-cluster-manager will create a base
   FRRConfiguration CR in the `frr-k8s-system` namespace, and a
   RouteAdvertisements CR that advertises the default network. The generated
   FRRConfiguration will configure BGP speakers on each node to advertise the

>    pod subnets and set up a full-mesh BGP topology where each node peers with

Reviewer 3: nit: full-mesh iBGP topology

just to be explicit

            toReceive:
              allowed:
                prefixes:
                - ge: 24
                  le: 24
                  prefix: 10.128.0.0/16
    ```

  The following steps are the same as steps 5 to 7 in the previous section.

#### Create a ClusterUserDefinedNetwork in No-overlay Mode with Unmanaged Routing

Here is a configuration example:

1. The frr-k8s pods shall be deployed on each node.
2. A cluster admin wants to enable no-overlay mode for the blue network by
   creating the following ClusterUserDefinedNetwork CR.

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: ClusterUserDefinedNetwork
    metadata:
      name: blue
      labels:
        network: blue
    spec:
      namespaceSelector:
        matchExpressions:
        - key: kubernetes.io/metadata.name
          operator: In
          values: ["ns1", "ns2"]
      network:
        topology: Layer3
        layer3:
          role: Primary
          # The UDN MTU shall not be larger than the provider network's MTU.
          mtu: 1500
          subnets:
          - cidr: 10.10.0.0/16
            hostSubnet: 24
        transport: "NoOverlay"
        noOverlayOptions:
          outboundSNAT: "Disable"

>           routing: "Unmanaged"

Author: Yes

        asn: 64512
        disableMP: true
        toReceive:
          allowed:
            mode: all
      - address: 192.168.111.21
        asn: 64512
        disableMP: true
        toReceive:
          allowed:
            mode: all
      - address: 192.168.111.22
        asn: 64512
        disableMP: true
        toReceive:
          allowed:
            mode: all
```

##### External Route Reflectors for Larger Clusters

In a large cluster, a full-mesh BGP setup leads to more CPU and memory
consumption on the nodes. Instead of every node peering with every other node,
nodes peer only with external BGP route reflectors. This significantly reduces
the number of BGP sessions each node needs to maintain, improving scalability.

```mermaid
graph TD
    subgraph AS 64512 [ AS 64512 ]
        direction LR
        subgraph "Route Reflector"
            RR1(Route Reflector 1)
        end
        subgraph "Kubernetes Nodes"
            NodeA(Node A)
            NodeB(Node B)
            NodeC(Node C)
        end
    end
    NodeA <-- iBGP --> RR1
    NodeB <-- iBGP --> RR1
    NodeC <-- iBGP --> RR1
```

### Feature Compatibility

#### Multiple External Gateways (MEG)

> MEG cannot work with advertised networks, therefore MEG cannot work with

Reviewer 3: hmm I forgot I guess that we excluded MEG support with CDN + BGP. I guess you can ignore my earlier comments about MEG and we can handle feature support in a later OKEP.

networks operating in the no-overlay mode either.

#### Egress IP

> Not supported.

Author: Yeah, it cannot work if the nodes are not using each other as nexthop. So even if we implement EIP with no-overlay mode, there is still a limitation.

* The `topology` flag specifies the BGP topology to use for the cluster internal
  BGP fabric, with supported values of `full-mesh` initially, more topology
  options can be added in the future. It's optional.

  This flag is required when any no-overlay network is configured with `managed`
  routing. When set, the `ovnkube-cluster-manager` generates FRRConfigurations
  to configure the BGP topology for the cluster.

  If set to `full-mesh`, ovnkube-cluster-manager will generate FRRConfigurations
  for all nodes to peer with each other in a full mesh. This topology is simpler
  to configure but less scalable for large clusters.

* The `as-number` flag is used when no-overlay networks are configured with
  managed routing. It is shared by both the cluster default network and CUDNs.
  It specifies the AS number to be used by the BGP speakers on each node for its
  default VRF. It's optional. The default AS number is `64512` if not specified.

For CUDNs, no-overlay mode is configured via their respective CRs. Due to a BGP
limitation, CUDNs with overlapping pod subnets cannot be advertised to the same
VRF. To enable no-overlay mode for these networks, a VRF-lite configuration is

> required.

Author: I'll add this to the test plan.

Here is an example of how to configure the iBGP neighbors on FRR routers:

```text
  neighbor 172.18.0.2 activate
  neighbor 172.18.0.2 route-reflector-client
  neighbor 172.18.0.2 next-hop-self force
```

### Feature Compatibility

#### Multiple External Gateways (MEG)

MEG cannot work with advertised networks, therefore MEG cannot work with
networks operating in the no-overlay mode either.

#### Egress IP

Not supported.

#### Services

MetalLB will still be used in order to advertise services across the BGP fabric.

#### Egress Service

Not supported.

#### Egress Firewall

Full support.

#### Egress QoS

Full support.

#### Network Policy/ANP

Full support.

#### IPsec

Not supported.

#### Multicast

Not supported.

OVN does not support routing multicast traffic to external networks. OVN logical
routers only support forwarding multicast traffic between logical switches. It
cannot exchange PIM messages with external routers. Additionally, FRR-K8s does
not currently support configuring the FRR multicast functions (IGMP/MLD, PIM,
etc.). Therefore, multicast is not supported in no-overlay mode.

### Testing Details

* E2E Testing Details

>   The E2E tests shall cover the following combinations:

Reviewer 4: > CDN geneve <-> no-overlay transition is not supported.

hmm wouldn't this be the first thing we don't allow to reconfigure on restart?

Not sure if that's reasonable but @ricky-rav something to account for in your downstream enhancement,

network to integrate seamlessly with existing BGP-capable networks and achieve
maximum network performance.

### Story 3: Deploy networks in the no-overlay mode without external BGP routers

As a cluster admin, I want to use no-overlay mode for intra-cluster traffic
without advertising pod networks to the external network or depending on
external BGP routers.

## Proposed Solution

This solution leverages the existing BGP feature to advertise each node's pod
subnet via the CNCF project [FRR-K8s](https://github.com/metallb/frr-k8s). FRR
routers managed by FRR-k8s then propagate these routes throughout the provider
network, enabling direct pod-to-pod traffic without an overlay. A single cluster
can simultaneously have networks in both overlay and no-overlay modes.

When no-overlay mode is enabled for a network:

* By default, when BGP advertisement is enabled, the BGP feature disables SNAT
  for pod egress traffic leaving nodes. For users who only want to have
  no-overlay mode but do not want to expose pod IPs externally, this behavior is
  not ideal. To address this, users will be able to enable SNAT for
  pod-to-external traffic. This will be configured via a new `outboundSNAT`

>   field in the `ClusterUserDefinedNetwork` CRD for CUDNs, and the

Reviewer 5: Just a note for my future self: That CUDN in itself when created is always going to SNAT pod to external to nodeIP.
What introduces the no-SNAT concept is the RA CRD, however the field to override that is being added via CUDN CRD.
Reasoning for this is because we want the granularity for this per network I guess. The network overlay itself definitely belongs to the network API, just the outboundSNAT is slightly misplaced but I am ok with the current proposal.

no-overlay mode, this key is set to `noOverlay`. Additionally, the `outboundSNAT`
key will be included to indicate the SNAT behavior for outbound traffic from pods.
This information can then be passed into the NetInfo object and utilized by the
route import and gateway controller of ovnkube-controller.

Note: Users must not set these fields manually; OVN-Kubernetes manages them.

An example OVN-Kubernetes NAD may look like:

```yaml
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: l3-network
  namespace: default
spec:
  config: |
    {
            "cniVersion": "0.3.1",
            "name": "l3-network",
            "type": "ovn-k8s-cni-overlay",
            "topology":"layer3",
            "mtu": 1500,
            "netAttachDefName": "default/l3-network",
            "role": "primary",
            "transport": "noOverlay",
            "outboundSNAT": "disabled"
    }
```

##### Host Routing Tables

To [maintain isolation between
UDNs](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5140), node BGP
speakers do not accept routes to pod subnets by default. However, to route

> east-west traffic in no-overlay mode, these routes must be imported to the host

Reviewer 5: super nit: I'd clarify that we would thus import remote node pod routes *if no-overlay is enabled for a given network* and otherwise behaviour is unchanged.

leveraging ovn-kubernetes support for BGP, we want to provide a way for users to
enable a no-overlay mode, which would disable Geneve encapsulation and use
direct routing between nodes for east/west traffic on selected networks.

Many environments, particularly on-premise deployments or those with dedicated
networking infrastructure, prefer to utilize the underlying physical network's
routing capabilities directly. This "no-overlay" approach can offer several
benefits:

* Improved Performance: Eliminates encapsulation/decapsulation overhead,
  potentially leading to lower latency and higher throughput for inter-pod
  communication.
* Simplified Troubleshooting: Traffic paths are more transparent as they align
  with the physical network's routing tables, simplifying debugging and network
  visibility.
* Leverage Existing Network Infrastructure: Integrates more seamlessly with
  existing BGP-capable network devices, allowing for direct routing to pod IPs.
* Reduced Resource Consumption: Less CPU cycles spent on
  encapsulation/decapsulation.

## Goals

* Support no-overlay mode for the default network
* Support no-overlay mode for layer-3 Cluster User Defined Networks (CUDNs) in

>   the VRF-lite mode .

Reviewer 6: Here montioned VRF-lite mode,  How about default VRF mode for CUDN?

Reviewer 6: OK,  please ignore this,  found the details explanation in the following part.

* A cluster can have networks operating in overlay and no-overlay modes
  simultaneously.
* Use the route advertisements feature to exchange routes to node subnets across
  the cluster.
* Allow direct pod-to-pod communication without any overlay transport for
  West-East traffic.
* Ensure East-West network policies are still enforced by OVN.
* Maintain compatibility with existing OVN-Kubernetes features where applicable
  (e.g., Services, NetworkPolicy, EgressIP, EgressFirewall, AdminNetworkPolicy).
* Compatible with both local gateway and shared gateway modes.
  
## Future Goals

* When OVN-Kubernetes supports BGP for UDNs, extend the UDN API to enable
  no-overlay mode on the selected network

> * Support toggling no-overlay mode on/off for an existing network.

Reviewer 6: Support togging no-overlay mode on/off runtime by patch the CUDN resource, right?

Author: The default network and UDNs. We can remove the API restriction if this use case makes sense.

### Story 3: enable no-overlay mode without an external BGP router

As a cluster admin, I want to avoid all transport for traffic in the cluster,
but I don't want to deploy an external BGP router for the OCP cluster.

## Proposed Solution

The core idea is to leverage the existing Route Advertisement feature to
advertise the IP subnet allocated to each node (which contains the IPs of pods
running on that node) to an internal/external BGP route reflector. The BGP route
reflector will then populate these routes throughout the physical network,
allowing other nodes to directly route traffic to pod IPs without needing an
overlay. Within a cluster, networks can operate in different overlay modes.
Users can have overlay and no-overlay networks simultaneously.

When the no-overlay mode is enabled for a network:

* For north-south traffic, we will follow the implementation of the Route
  Advertisement feature, the traffic from pods will no longer be SNATed on
  egress. In the route advertisement feature, SNAT is disabled for all pod
  egress traffic that leaves the node, regardless of whether the destination
  route is BGP learnt or not. For users who only want to have no-overlay mode

>   but not want to expose podIPs to external, this behavior is not ideal. Instead

Author: Why not. No-overlay is for east-west, exposing pod IP to external is for north-south.

  of disabling SNAT for all the pod-to-external traffic, OVN-K shall adjust its
  SNAT behavior so that pod IPs are exposed for traffic routed via BGP, but SNAT

>   is applied for non-BGP-routed egress traffic.

Author: Actually, we've decided to make this change to the BGP implementation. It's for the BGP feature, not only for no-overlay mode. Once that change is done, we can remove these sentences from this doc.

* For east-west traffic, intra-node traffic (pod-to-pod, pod-to-clusterIP and
  pod-to-host) remains unchanged, while cross-node traffic will follow the same
  path as north-south traffic.

### API Details

We introduce a new `layer3.encapsulation` field to be added to the Spec of the
ClusterUserDefinedNetwork (CUDN) CRD. This
new field will enable control over the network transport behavior with the
following options:

1. **Encapsulation Configuration**:
   * `encapsulation`: specifies the transport method (default: `Geneve`)
   * Supported values: `Geneve` (default) or `None` (no overlay)

The `spec` field of a ClusterUserDefinedNetwork CR is immutable. Therefore, the
encapsulation configuration cannot be changed once a ClusterUserDefinedNetwork

> CR is created.

Author: This is the current implementation. https://github.com/ovn-kubernetes/ovn-kubernetes/blob/682ffaf5d4550fd5dac1624ef53de7ee5d82416f/dist/templates/k8s.ovn.org_clusteruserdefinednetworks.yaml.j2#L537

> CR is created.

Reviewer 2: Yeah, that's true. If this aligns with the expectations for this feature, all good :)

#### Example of a layer-3 CUDN with no-overlay mode enabled

A layer-3 CUDN that enables no-overlay mode should look like this:

```yaml
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: my-cudn
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name
      operator: In
      values: ["red", "blue"]
  network:
    topology: Layer3
    layer3:
      role: Primary
      mtu: 1500
      subnets:
      - cidr: 10.10.0.0/16
        hostSubnet: 24
      encapsulation: None
```

#### Example of the default network with no-overlay mode enabled

For the cluster default network, the cluster admin shall create a

> ClusterUserDefinedNetwork CR with the reserved name `default` in the namespace where

Reviewer 6: is that possible enable no-overlay node for default network by CNO like `no-overlay-default-network: enable` and let CNO to create a default ClusterUserDefinedNetwork CR ?

    layer3:
      role: Primary
      mtu: 1500
      subnets:
      - cidr: 10.244.0.0/16
        hostSubnet: 24
      encapsulation: None
```

### Implementation Details

No-overlay mode is only available when OVNK interconnect is enabled, since it
relays on the route advertisement feature, which only works with interconnect
mode.

A network that is configured to use no-overlay mode needs a slightly modified
OVN topology. It is useful to remember that in today's interconnect architecture
a network is able to span all cluster nodes thanks to the cross-node logical
links that OVN establishes between the transit switch and the ovn-cluster-router
instances on each node. That is the logical representation of the OVN network
and in practice these links are implemented through Geneve tunnels between every
node in the cluster. For a network to skip the transport step and route pod
traffic directly over the underlay network, we need to change the logical
network topology of this network: its ovn-cluster-router won't connect to the
transit switch anymore and all ease-west traffic will be naturally routed in the
same path as north-south traffic through the provider network.

> In Shared Gateway Mode, at the OVN ClusterRouter, the pod to remote pod egress

Reviewer 6: OK, So the EIP pods egress outside, still overlay to egress node

> In Shared Gateway Mode, at the OVN ClusterRouter, the pod to remote pod egress

Author: I may have underestimated the complexity of supporting EIP in no-overlay mode. Using the overlay network to forward traffic from a pod to the egress node may not be the best approach.

GatewayRouter, it will be forwarded the nexthop node according to routes learnt
from BGP.

In Local Gateway Mode, the pod to remote pod traffic egress traffic will enter
the host VRF via the ovn-k8s-mpX interface. Then it will be forwarded to remote
nodes according to the host routes of the VRF.

The following diagram shows the OVN topology of a cluster with three networks:

* a CUDN (in green) that uses Geneve transport (default behavior)
* a default network (in gray) that enables no-overlay mode
* a CUDN (in blue) that enables no-overlay mode

![No-overlay topology](../images/no_overlay_topology.jpg)

#### No SNAT in no-overlay mode for egress traffic

The BGP feature in ovn-kubernetes only disables SNAT for traffic that is
destined to external networks, not for traffic that is destined to other pods in
the cluster.

In no-overlay mode, we want to fully leverage BGP and avoid the SNAT step on the
gateway router for intra-cluster traffic as well. Pod-to-pod traffic within the
same network will hit the BGP-learned routes to reach its destination node,
without being encapsulated in a Geneve tunnel and without being SNATed to the
node IP.

> The only exception is for pod to other host traffic, to ensure nodePort services

Reviewer 2: Uhm here I don't follow. Are we keeping the SNAT so that we can allow traffic to nodeports on different nodes, or are we keeping the SNAT so that users won't expose pod IPs to the outside?

Author: We keep the SNAT only for pod-to-remote_host traffic.

can be accessed cross networks. The egress traffic will be SNATed to the node IP
before leaving the node.

#### Isolation between UDNs with no-overlay mode enabled

The [existing UDN isolation
mechanism](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5186/commits/8f6e7d30ee5f4926a21e2de75488aad80344814b)
will still be in place. Pods in different UDNs will be isolated from each other
and from the default network by means of the existing OVN ACLs on every node
logical switch: ACLs are enforced on outgoing traffic at each advertised UDN
switch, verifying whether both source and destination IPs belong to the same
advertised UDN. If the destination IP does not come from the same UDN subnet as
the source IP, traffic is dropped. The no-overlay mode will not affect the
isolation between UDNs or between UDNs and the default network.

### Workflow

#### Enable No-overlay Mode for the Default Network

No-overlay mode for the default network must be enabled before OVN-Kubernetes
pods start. Therefore, the following configuration manifests shall be created

> before deploying the OVN-Kubernetes pods.

Reviewer 6: so is that mean we need to re-created the ovn-kubernetes pods after creating the CR on existing cluster to enable the default network no-overlay?

Author: No, the CUDN CR shall be created before ovn-k pods start.

Reviewer 2: I didn't get this either. So can the default network be configured to use no-overlay mode only at installation time? Or are we telling users to kill ovn-k pods after creating the CUDN CR for the default network? I'm not sure if either case is ideal :)

1. The cluster admin enables no-overlay mode by creating the well-known
   `default` UDN CR with `spec.layer3.encapsulation` set to `None`.

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: ClusterUserDefinedNetwork
    metadata:
      # 'default' is the reserved name to represent the cluster default network
      name: default
    spec:
      # use an empty selector to select all namespaces

>       namespaceSelector: {}

Reviewer 2: Should we invalidate all other values for `namespaceSelector`?

Author: Maybe we can add a validation rule for this.

            disableMP: true
            toReceive:
              allowed:
                mode: filtered
    ```

1. OVN-Kubernetes will generate a following FRRConfiguration for each node.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:
      name: route-generated-blue
      namespace: frr-k8s-system
    spec:
      bgp:
        routers:
        - asn: 64512
          neighbors:
          - address: 10.89.0.37
            asn: 64512
            disableMP: true
            toAdvertise:
              allowed:
                prefixes:
                # the node subnet of this node
                - 10.244.1.0/24
          prefixes:
          - 10.244.1.0/24
    ```

#### Create A ClusterUserDefinedNetwork in No-overlay Mode

For user defined networks, to ensure the traffic isolation, no-overlay mode can
only be enabled when the route advertisement is configured in a VFR-Lite mode.
The cross node east-west traffic in this user defined network will be isolated
the by the VRFs in each node and the VLANs of the provider network. VRF-Lite is
only available when using Local Gateway Mode, thus no-overlay mode user defined

> networks is also only available when using Local Gateway Mode.

Author: I want to ensure the network isolation is implemented end-to-end (not only within a node but also in the provider network), similar to what we used to have in overlay mode. The node overlay mode is for east-west traffic; mixing routes for different UDNs in a single VRF could introduce a parental security vulnerability.

Here's a configuration example:

1. A user want to enable no-overlay mode for the blue network. The user has to

>    do all host modifications necessary via NMState to enslave an VLAN interface

Reviewer 2: ```suggestion
   do all host modifications necessary via NMState to enslave a VLAN interface
```

          - key: kubernetes.io/metadata.name
            operator: In
            values: ["ns1", "ns2"]
        network:
          topology: Layer3
          layer3:
            role: Primary
            # The UDN MTU shall not be larger than the provider network's MTU.
            mtu: 1500
            subnets:
            - cidr: 10.10.0.0/16
              hostSubnet: 24
            encapsulation: None
    ```

1. The user has created an FRRConfiguration CR to peer with a external BGP
   router `182.18.0.5` on the blue VLANs.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:
      name: vpn-blue
      namespace: frr-k8s-system
      labels:
        network: vpn-blue
    spec:
      bgp:
        routers:
        - asn: 64512
          vrf: blue
          neighbors:
          - address: 182.18.0.5
            asn: 64512
            disableMP: true
            holdTime: 1m30s
            keepaliveTime: 30s
            passwordSecret: {}
            port: 179
            toAdvertise:
              allowed:
                mode: filtered
            toReceive:
              allowed:
                mode: filtered
    ```

> 1. The user have to advertise this pod network of the CUDN by BGP. The targetVRF

Reviewer 2: ```suggestion
1. The user has to advertise this pod network of the CUDN by BGP. The targetVRF
```

      # nodeSelector must be empty, since we don't support a network in a overlay and no-overlay hybrid mode.
      nodeSelector: {}
      frrConfigurationSelector:
        network: vpn-blue
      networkSelectors:
      - networkSelectionType: ClusterUserDefineNetwork
        clusterUserDefinedNetworkSelector:
          networkSelector:
            matchLabels:
              # Select one CUDN to ensure the cross UDN isolation
              advertise: blue
      advertisements:
      - PodNetwork
    ```

1. OVN-Kubernetes will generate a following FRRConfiguration for each node.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:
      name: route-generated-blue
      namespace: frr-k8s-system
    spec:
      bgp:
        routers:
        - asn: 64512
          neighbors:
          - address: 182.18.0.5
            disableMP: true
            asn: 64512
            toAdvertise:
              allowed:
                prefixes:
                # the node subnet of this node
                - 10.10.1.0/24
          vrf: blue
          prefixes:
          - 10.0.1.0/24
    ```

### Deployment Consideration

#### BGP Topology

> The deployment mode will be decided by the FRRConfiguration CR that is selected

Reviewer 2: To stress that it's up to the cluster admin:
```suggestion
The deployment mode will be decided by the FRRConfiguration CR that the user creates
```

BTW should we replace `user` with `cluster admin` all throughout the enhancement?

no-overlay mode without relying on external BGP routes.

##### External Route Reflectors for Larger Clusters

In a large cluster, a full-mesh BGP setup leads to more CPU and memory
consumption on the nodes. Instead of every node peering with every other node,
nodes peer only with external BGP route reflectors. This significantly reduces
the number of BGP sessions each individual node needs to maintain, improving
scalability.

##### Internal Route Reflectors

A FRR instances can also function as a BGP route reflectors, technically we can
use internal route reflectors to build a BGP network across the cluster.
However, the frr-k8s which is used for integrating OVN-K with FRR does not
currently support the configuration of FRR as a route reflector, therefore this
deployment mode will not be supported.

### Feature Compatibility

#### Multiple External Gateways (MEG)

The same as the route advertisement feature.

#### Egress IP

Full support

#### Services

The same as the route advertisement feature.

#### Egress Service

Full support.

#### Egress Firewall

Full support.

#### Egress QoS

Full Support.

#### Network Policy/ANP

Full Support.

#### IPsec

> Not Support.

Reviewer 6: so on ipsec cluster, there will `not supported` message shown when applying the no overlay CR?

maximizing throughput. Users who intend to enable BGP on their clusters can
indeed leverage BGP-learned routes to achieve this. The goal is to provide users
with an API to enable or disable no-overlay mode on selected networks (default
or user-defined), allowing traffic to skip Geneve encapsualtion (i.e. the
overlay network) and simply make use of the learned routes in the underlay or
provider network for inter-node communication.

However, BGP is not the only option of exchanging the east-west routing
information within a cluster. In the future, the no-overlay mode may support
other approaches.

## User-Stories/Use-Cases

### Story 1: enable no-overlay mode for the default network

As a cluster admin, I want to avoid all transport for traffic in the default
network to integrate seamlessly with existing BGP-capable network, achieve
maximum network performance and simplify troubleshooting.

### Story 2: enable no-overlay mode for a CUDN

As a cluster admin, I want to avoid all transport for traffic in a CUDN to
integrate seamlessly with existing BGP-capable network, achieve maximum network
performance and simplify troubleshooting.

> ### Story 3: enable no-overlay mode without an external BGP router

Reviewer 4: Got it now.  I consider it default behavior (although it isn't), let's call it loose isolation mode.

Right, it is a problem and I don't have a straight answer right now.

First we would need to make an assumption of what is the next hop for that traffic (if we don't want it to be directly the cluster node). Sans an extra API, we would have to assume it is the default gateway.

For SGW mode, we have the option of not importing other networks routes into the GR.

I don't think we have an equivalent alternative in LGW mode. I thought about FRR facilities for blacklisting/rewriting imported/received routes, because I think that is the best place to solve this, but couldn't find/come up with anything so far. 

So the only thing I can think of right now is use the current strict isolation nft rules to mark packets instead of dropping them, and then use an ip rule based on this mark to sent traffic to default gateway.

> ### Story 3: enable no-overlay mode without an external BGP router

Reviewer 3: @pliurh I think we still need to figure this one out

> ### Story 3: enable no-overlay mode without an external BGP router

Reviewer 3: yes, sorry, I had it backwards. @pliurh can you make sure we have a section in the enhancement covering this?

When the no-overlay mode is enabled for a network:

* For north-south traffic, we will follow the implementation of the Route
  Advertisement feature, the traffic from pods will no longer be SNATed on
  egress. In the route advertisement feature, SNAT is disabled for all pod
  egress traffic that leaves the node, regardless of whether the destination
  route is BGP learnt or not. For users who only want to have no-overlay mode
  but not want to expose podIPs to external, this behavior is not ideal. Instead
  of disabling SNAT for all the pod-to-external traffic, OVN-K shall adjust its
  SNAT behavior so that pod IPs are exposed for traffic routed via BGP, but SNAT
  is applied for non-BGP-routed egress traffic.

* For east-west traffic, intra-node traffic (pod-to-pod, pod-to-clusterIP and
  pod-to-host) remains unchanged, while cross-node traffic will follow the same
  path as north-south traffic.

### API Details

We introduce a new `layer3.encapsulation` field to be added to the Spec of the
ClusterUserDefinedNetwork (CUDN) CRD. This
new field will enable control over the network encapsualtion behavior with the
following options:

> 1. **Encapsulation Configuration**:

Author: Using `VXLAN` for EVPN could be misleading. Since OVN itself also supports using vxlan as overlay protocol. Maybe, we could use `EVPN_VXLAN` or `EVPN_MPLS` instead.

> 1. **Encapsulation Configuration**:

Reviewer 4: If the concern is overwriting a user provided value (valid concern in my opinion), then we can require the user to set that field to the required value instead, otherwise we don't reconcile the EVPN CRD. 

In terms of what value should that be: perhaps `None` if the network is going to be configured in the same way as no overlay mode without EVPN, or perhaps we can require the user to set special value `EVPN_VXLAN` even if the network is not configured differently as a form of mutual acknowledgement or coordination, example: with a special value maybe we know we need to wait for something else to happen to reconfigure the network.

> 1. **Encapsulation Configuration**:

Reviewer 3: @pliurh I think we still need to figure this one out as well

> 1. **Encapsulation Configuration**:

Reviewer 3: I think the problem with that is if someone selects multiple EVPNs...hmm, maybe this is a flaw in my EVPN CRD as well. Someone cannot select more than 1 CUDN for MAC-VRF (L2 domain) that wouldn't make any sense connecting multiple L2 domains to the same L2 domain. 

They could select more than 1 CUDN for an IP-VRF though, since that is pure routing, as long as they don't have overlapping subnets.

Maybe your suggestion is a very good one. Then I can have several CUDNs select the same EVPN CRD, and if they are layer 3, they will just use the ip-vrf ID specified there. If the CUDN specifies a layer 2 and evpnSelector, and that EVPN CR already has another layer2 network configured for it, then it reports error status to the CUDN. That will tell the user that CUDN network failed, rather than sending that status in the EVPN CR.

I like this approach! @jcaamano wdyt?

   * `encapsulation`: specifies the encapsualtion method (default: `Geneve`)
   * Supported values: `Geneve` (default) or `None` (no overlay)

The `spec` field of a ClusterUserDefinedNetwork CR is immutable. Therefore, the
encapsulation configuration cannot be changed once a ClusterUserDefinedNetwork
CR is created.

#### Example of a layer-3 CUDN with no-overlay mode enabled

A layer-3 CUDN that enables no-overlay mode should look like this:

```yaml
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: my-cudn
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name
      operator: In
      values: ["red", "blue"]
  network:
    topology: Layer3
    layer3:
      role: Primary
      mtu: 1500
      subnets:
      - cidr: 10.10.0.0/16
        hostSubnet: 24

>       encapsulation: None

Author: UDN support is a future goal. We can validate it in the cluster manager and set a `status.condition` in the UDN CR to indicate that this configuration is not supported yet.

```

#### Example of the default network with no-overlay mode enabled

For the cluster default network, the cluster admin shall create a
ClusterUserDefinedNetwork CR with the reserved name `default` and set

> encapsulation to `None` to enable no-overlay mode before deploying

Author: I've also considered this option. I considered 2 points: 1) implementing a unified API for all ovn-k networks; and 2) planning for future support of migration between different encapsulation methods.

> encapsulation to `None` to enable no-overlay mode before deploying

Reviewer 4: Additionally, if we are going to use a CUDN for this, then we should describe what happens when:
- the `default` CUDN is created after OVN-Kubernetes has already started
- the `default` CUDN is deleted after OVN-Kubernetes has already started
- the `default` CUDN is created after OVN-Kubernetes has already started but while being down

So, agreeing with @trozet here, using a CUDN when it is not supposed to be a day-2 thing seems like an extra complication.

> encapsulation to `None` to enable no-overlay mode before deploying

Author: We need to decide whether to support changing the encapsulation for the default network during day-2 operations. If the answer is no, then I agree we shall use a flag instead of the reserved CR.

OVN-Kubernetes. For this CR, OVN-K shall not create the corresponding

> NetworkAttachementDefinition CR. The subnet field must aligned with the value in

Author: Sure, but the NAD doesn't have to be derived from the `default` CUDN unless that's our intention.

> NetworkAttachementDefinition CR. The subnet field must aligned with the value in

Reviewer 4: I don't know what role the NAD (any NAD) has on this enhancement. I think it doesn't have any role or at least it is not mentioned.

advertised UDN. If the destination IP does not come from the same UDN subnet as
the source IP, traffic is dropped. The no-overlay mode will not affect the
isolation between UDNs or between UDNs and the default network.

### Workflow

#### Enable No-overlay Mode for the Default Network

No-overlay mode for the default network must be enabled before OVN-Kubernetes
pods start. Therefore, the following configuration manifests shall be created
before deploying the OVN-Kubernetes pods.

1. The cluster admin enables no-overlay mode by creating the well-known
   `default` UDN CR with `spec.layer3.encapsulation` set to `None`.

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: ClusterUserDefinedNetwork
    metadata:
      # 'default' is the reserved name to represent the cluster default network
      name: default
    spec:
      # use an empty selector to select all namespaces
      namespaceSelector: {}
      network:
        topology: Layer3
        layer3:
          role: Primary
          mtu: 1500
          subnets:
          - cidr: 10.244.0.0/16
            hostSubnet: 24
          encapsulation: None
    ```

1. OVN-kubernetes cluster manager watches the `default` UDN CR. If

>    `spec.layer3.encapsulation` equals `None`, cluster manager will create a

Reviewer 4: Just noting that this could be done on a different RA and configuration would be aggregated.

    metadata:
      name: receive-filtered
      namespace: frr-k8s-system
      labels:
        network: default
    spec:
      bgp:
        routers:
        - asn: 64512
          neighbors:
          # An external BGP neighbor acts as a route reflector
          - address: 10.89.0.37
            asn: 64512
            disableMP: true
            toReceive:
              allowed:
                mode: filtered
    ```

1. OVN-Kubernetes will generate a following FRRConfiguration for each node.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:
      name: route-generated-blue
      namespace: frr-k8s-system
    spec:
      bgp:
        routers:
        - asn: 64512
          neighbors:
          - address: 10.89.0.37
            asn: 64512
            disableMP: true
            toAdvertise:
              allowed:
                prefixes:
                # the node subnet of this node
                - 10.244.1.0/24
          prefixes:
          - 10.244.1.0/24
    ```

#### Create A ClusterUserDefinedNetwork in No-overlay Mode

For user defined networks, to ensure the traffic isolation, no-overlay mode can

> only be enabled when the route advertisement is configured in a VFR-Lite mode.

Author: Technically, it can be done. But it will break the end-to-end isolation. It could introduce a potential security vulnerability. No-overlay mode should maintain the same level of isolation as overlay mode. But it's just my thought. I'm okay with it if we want the user to decide.

      - networkSelectionType: ClusterUserDefineNetwork
        clusterUserDefinedNetworkSelector:
          networkSelector:
            matchLabels:
              # Select one CUDN to ensure the cross UDN isolation
              advertise: blue
      advertisements:
      - PodNetwork
    ```

1. OVN-Kubernetes will generate a following FRRConfiguration for each node.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:
      name: route-generated-blue
      namespace: frr-k8s-system
    spec:
      bgp:
        routers:
        - asn: 64512
          neighbors:
          - address: 182.18.0.5
            disableMP: true
            asn: 64512
            toAdvertise:
              allowed:
                prefixes:
                # the node subnet of this node
                - 10.10.1.0/24
          vrf: blue
          prefixes:
          - 10.0.1.0/24
    ```

### Deployment Consideration

#### BGP Topology

The deployment mode will be decided by the FRRConfiguration CR that is created
for the no-overlay network. There will be no new field added to the
FRRConfiguration.

##### Full-Mesh iBGP across the Cluster

> The FRR instance on each node maintains a full mesh BGP peer relationship with

Author: This can be done by adding all the node router IDs into the same FrrConfiguration. We don't need to make any changes to the BGP implementation. It already works today.

all other nodes across the cluster. In this mode, users can enable the
no-overlay mode without relying on external BGP routes.

> ##### External Route Reflectors for Larger Clusters

Author: Do you mean using eBGP within the cluster? Is this a valid use case? Also, BGP implementation doesn't support such a configuration. Routes to a podNetwork can only be advertised together.
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/682ffaf5d4550fd5dac1624ef53de7ee5d82416f/dist/templates/k8s.ovn.org_routeadvertisements.yaml.j2#L546

> ##### External Route Reflectors for Larger Clusters

Reviewer 3: You don't need to have iBGP peering relationships between the nodes. You can use eBGP with multihop +loopback. You put each node in its own AS and then you can either peer full mesh style, or segment peering how you want to. This is useful in a cluster with a large amount of worker nodes, where you cant afford to do full mesh.

> ##### External Route Reflectors for Larger Clusters

Author: I see. However, I am uncertain whether our current BGP implementation supports such a configuration. OVN-K shall generate the frrconfiguration CR for each node, using their respective original frrconfiguration CRs. @jcaamano WDYT?

> ##### External Route Reflectors for Larger Clusters

Reviewer 4: We already can generate frrconfiguration CR for each node from a frrconfiguration CR for each node. So I don't see anything that prevents this configuration right now.
But this, as many others, would be a configuration combination that we are not doing integration testing for. Given the amount of alternatives, we need to either restrict the set that we support in some way, and add to them over time, or trust they just work because we don't see a reason they wouldn't. 
I see the extra possibilities of eBGP but I don't see how it would be preferred over a route reflector.
I don't have anything against mentioning it as an alternative. But if we think that we need to act on it in any other way, then we need to be more explicit.

> ##### External Route Reflectors for Larger Clusters

Reviewer 3: It would be preferred over a route reflector where:
1. There is no external bgp speaker.  Then something needs to designate at least 2 RRs (for HA) and now you need  BGP peering sessions from all nodes to both. Managing the RRs might become cumbersome. With eBGP you can break up large chunks of nodes into AS and then use multihop and build out the topology as you see fit. There is more flexibility.
2. There is an external bgp speaker, but they do not want to use iBGP. A lot of users will not want to use iBGP. I think it will be more common to use eBGP. It provides clear AS boundaries and allows for some extra functionality (AS path prepend, MED, filtering on as path). Also the next hop with iBGP will not set to be the worker node IP, unless it is peered "with next-hop-self".

> ##### External Route Reflectors for Larger Clusters

Author: Yeah, this is a valid use case. I added an eBGP topology example to the document. PTAL.

## Non-Goals

* This enhancement does not aim to change the default behavior of
  OVN-Kubernetes, which will continue to use Geneve encapsualtion for the
  default network and any user-defined networks unless no-overlay mode is
  explicitly enabled.
* This enhancement does not aim to change the existing CUDN/UDN isolation
  mechanism. CUDNs/UDNs will continue to be isolated from each other and from
  the default network by means of OVN ACLs on every node logical switch.
* This enhancement does not aim to change the existing BGP configuration or
  behavior. The user must ensure that the BGP configuration is correctly set up
  to support no-overlay mode.
* This enhancement does not aim to change the existing CUDN lifecycle
  management. The user must ensure that the CUDN CRs are correctly managed
  according to the existing lifecycle management practices.
* This enhancement does not aim to implement the no-overlay mode with the
  centralized OVN architecture.
* This enhancement does not aim to implement the no-overlay mode for layer-2 or
  localnet type of networks.
* This enhancement does not aim to support an overlay and no-overlay hybrid

>   network.

Reviewer 4: BTW, I can really think a few reasons why Layer-2 support is more challenging. 

But what I really tried to say is that its better to write them down in the enhancement so we can comment on those if needed.

maximizing throughput. Users who intend to enable BGP on their clusters can
indeed leverage BGP-learned routes to achieve this. The goal is to provide users
with an API to enable or disable no-overlay mode on selected networks (default
or user-defined), allowing traffic to skip Geneve encapsualtion (i.e. the
overlay network) and simply make use of the learned routes in the underlay or
provider network for inter-node communication.

However, BGP is not the only option of exchanging the east-west routing
information within a cluster. In the future, the no-overlay mode may support
other approaches.

## User-Stories/Use-Cases

### Story 1: enable no-overlay mode for the default network

As a cluster admin, I want to avoid all encapsulation for traffic in the default
network to integrate seamlessly with existing BGP-capable network, achieve
maximum network performance and simplify troubleshooting.

### Story 2: enable no-overlay mode for a CUDN

As a cluster admin, I want to avoid all encapsulation for traffic in a CUDN to
integrate seamlessly with existing BGP-capable network, achieve maximum network
performance and simplify troubleshooting.

> ### Story 3: enable no-overlay mode without an external BGP router

Author: This is implemented by using a specific FrrConfiguration, which configures all nodes as BGP peers. It's already available with the current BGP implementation. I can give a configuration example for this setup.

      operator: In
      values: ["red", "blue"]
  network:
    topology: Layer3
    layer3:
      role: Primary
      mtu: 1500
      subnets:
      - cidr: 10.10.0.0/16
        hostSubnet: 24
      encapsulation: None
```

#### Example of the default network with no-overlay mode enabled

For the cluster default network, the cluster admin shall create a
ClusterUserDefinedNetwork CR with the reserved name `default` and set
encapsulation to `None` to enable no-overlay mode before deploying
OVN-Kubernetes. For this CR, OVN-K shall not create the corresponding
NetworkAttachementDefinition CR. The subnet field must be aligned with the value
in `cluster-subnets` parameter. The MTU shall not be larger than the node
interface MTU.

```yaml
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
Metadata:
  # 'default' is the reserved name to represent the cluster default network
  name: default
spec:
  # use an empty selector to select all namespaces
  namespaceSelector: {}
  network:
    topology: Layer3
    layer3:
      role: Primary
      mtu: 1500
      subnets:
      - cidr: 10.244.0.0/16
        hostSubnet: 24
      encapsulation: None
```

> ### Implementation Details

Reviewer 4: I am missing implementation details describing what OVNK components react to the new API and how.

Author: I will add.

network topology of this network: its ovn-cluster-router won't connect to the
transit switch anymore and all east-west traffic will be naturally routed in the
same path as north-south traffic through the provider network.

In Shared Gateway Mode, at the OVN ClusterRouter, the pod-to-remote_pod egress
traffic will be forwarded to the GatewayRouter via the join switch. Unlike the
current route advertisement behavior, the routes to remote node subnets will be
imported to the Gateway Router. Once the traffic reaches the GatewayRouter, it
will be forwarded the nexthop node according to routes learnt from BGP.

In Local Gateway Mode, the pod-to-remote_pod traffic egress traffic will enter
the host VRF via the ovn-k8s-mpX interface. Then it will be forwarded to remote
nodes according to the host routes of the VRF.

The following diagram shows the OVN topology of a cluster with three networks:

* a CUDN (in green) that uses Geneve encapsulation (default behavior)
* a default network (in gray) that enables no-overlay mode
* a CUDN (in blue) that enables no-overlay mode

![No-overlay topology](../images/no_overlay_topology.jpg)

> #### No SNAT in no-overlay mode for egress traffic

Reviewer 4: It is previously mentioned that SNAT will be kept in place for non-BGP next hops, but I am missing implementation details about this.

the source IP, traffic is dropped. The no-overlay mode will not affect the
isolation between UDNs or between UDNs and the default network.

### Workflow

#### Enable No-overlay Mode for the Default Network

No-overlay mode for the default network must be enabled before OVN-Kubernetes
pods start. Therefore, the following configuration manifests shall be created
before deploying the OVN-Kubernetes pods.

1. The cluster admin enables no-overlay mode by creating the well-known
   `default` UDN CR with `spec.layer3.encapsulation` set to `None`.

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: ClusterUserDefinedNetwork
    metadata:
      # 'default' is the reserved name to represent the cluster default network
      name: default
    spec:
      # use an empty selector to select all namespaces
      namespaceSelector: {}
      network:
        topology: Layer3
        layer3:
          role: Primary
          mtu: 1500
          subnets:
          - cidr: 10.244.0.0/16
            hostSubnet: 24
          encapsulation: None
    ```

1. OVN-kubernetes cluster manager watches the `default` CUDN CR. If
   `spec.layer3.encapsulation` equals `None`, cluster manager will create a

>    `RouteAdvertisement` CR which will advertise pod subnets in the default

Reviewer 4: So this RouteAdvertisements would be owned by ovnk. If this is the case then I think it is worthwhile to have some implementation details on how this is reconciled and how this status is reflected back to the CUDN crd status. I ask myself questions like: what happens if someone deletes this RouteAdvertisements?

Reviewer 4: On the other side, why does it it need to be created/owned by OVNK? Why isn't the user responsible for it?
Or, why don't we orchestrate no overlay mode for the default network from CNO? It might make sense to create the necessary FRRConfigurations, RouteAdvertisments and pass appropriate arguments to OVNK from there.

Reviewer 3: yeah I'm very wary of automatically creating this CR. I'd prefer for default network we just use a config flag. Unless you can tell us more about the plans for dynamically turning it on and off for default network in the future. I would argue it probably makes sense to force the user to re-roll out the ovnk daemonset, since this is like a one time config change that isnt going to change often.

Reviewer 4: To handle MTU issues, we have the MTU migration process. That would be enough for the default network. At some point we probably would need to find a way to loop in the UDNs in that MTU migration process, regardless of this enhancement.

   network:

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: RouteAdvertisements
    metadata:
      name: default
    spec:
      # nodeSelector must be empty, since we don't support a network in a overlay and no-overlay hybrid mode.
      nodeSelector: {}
      frrConfigurationSelector:
        matchLabels:

>           network: default

Reviewer 4: Missing details on the reasoning on what FRRConfiguration should be selected. Why match labels `network: default`?

Author: The no-overlay mode relies on BGP routers to exchange the routes to pod subnets. We use the selected FRRConfiguration to connect to the remote BGP router. The label is just an example.

          layer3:
            role: Primary
            # The UDN MTU shall not be larger than the provider network's MTU.
            mtu: 1500
            subnets:
            - cidr: 10.10.0.0/16
              hostSubnet: 24
            encapsulation: None
    ```

1. The cluster admin has created an FRRConfiguration CR to peer with a external
   BGP router `182.18.0.5` on the blue VLANs.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:
      name: vpn-blue
      namespace: frr-k8s-system
      labels:
        network: vpn-blue
    spec:
      bgp:
        routers:
        - asn: 64512
          vrf: blue
          neighbors:
          - address: 182.18.0.5
            asn: 64512
            disableMP: true
            holdTime: 1m30s
            keepaliveTime: 30s
            passwordSecret: {}
            port: 179
            toAdvertise:
              allowed:
                mode: filtered
            toReceive:
              allowed:
                mode: filtered
    ```

1. The cluster admin has to advertise this pod network of the CUDN by BGP. The
   targetVRF is set to auto, meaning routes to the podNetworks will only be

>    advertised within the corresponding VRF.

Reviewer 4: So if I am understanding this right, OVNK will create a RouteAdvertisements for the default network, but not for CUDNs?

Author: I've updated the steps. Users shall create the RA.

    participant Pod1 on Node A
    participant OVN Cluster Router
    participant OVN Gateway Router (GR)
    participant Underlay Physical Network
    participant Node B

    Pod1 on Node A->>OVN Cluster Router: Packet to remote Pod2 (10.128.2.11)
    OVN Cluster Router->>OVN Gateway Router (GR): Forward packet to GR via join switch
    Note over OVN Gateway Router (GR): BGP route lookup
    OVN Gateway Router (GR)->>OVN Gateway Router (GR): Match BGP route: 10.128.2.0/23 via 172.18.0.3 (Node B's IP)
    OVN Gateway Router (GR)->>Underlay Physical Network: Route packet out via br-ex
    Underlay Physical Network->>Node B: Packet arrives at Node B
    Note over Node B: Traffic is delivered to Pod2
```

To maintain SNAT for non-BGP-routed egress traffic, in the gateway router, the
default SNAT rule will be kept. The route import controller will add a Logical
Router Policy (LRP) to the gateway router for each BGP imported route to skip
the SNAT rule.

In the following example, `10.128.0.0/24` is the local nodeSubnet,
`174.16.1.0/24` is the CIDR of a BGP route.

```
100 ip4.src == 10.128.0.0/24 && ip4.dst == 174.16.1.0/24         outport                rtoe-GR_ovn-worker

> ```

Author: After some experiments, it appears that LRP cannot skip SNAT. I also tried to set the `exempted_ext_ips` flag in SNAT rules. It did not function as outlined in the OVN documentation. Do you know of any ways to implement this function in SGW mode?

> ```

Reviewer 4: Why not conditional SNAT as I mentioned before?

> ```

Reviewer 4: > > There could be another route with a longer prefix match that is not BGP.
> 
> I cannot think of a good solution for this case. The ideal approach is to bypass SNAT based on the nexthop IP. But OVN cannot do that.

Let's see what others think.

We could have a BGP for no-overlay only mode, where we just don't SNAT internal E/W traffic.

I guess it would not be complicated to calculate the routing space to come up with the real destinations that are BGP. However not sure such an effort should be scoped in the no-overlay epic if there are other acceptable alternatives.

    participant Host VRF/Kernel (on Node A)
    participant Underlay Physical Network
    participant Node B

    Pod1 on Node A->>OVN Cluster Router (on Node A): Packet to remote Pod2 (10.128.2.11)
    OVN Cluster Router (on Node A)->>Host VRF/Kernel (on Node A): Forward packet to host via ovn-k8s-mp0
    Note over Host VRF/Kernel (on Node A): BGP route lookup
    Host VRF/Kernel (on Node A)->>Host VRF/Kernel (on Node A): Match BGP route in host table: 10.128.2.0/23 via 172.18.0.3 (Node B's IP)
    Host VRF/Kernel (on Node A)->>Underlay Physical Network: Route packet out via eth0
    Underlay Physical Network->>Node B: Packet arrives at Node B
    Note over Node B: Traffic is delivered to Pod2
```

To maintain SNAT for non-BGP-routed egress traffic, the masquerade iptables rule
will be kept. The route import controller will insert a iptables rule for each
BGP imported route to skip the SNAT rule to the gateway router.

In the following example, `10.128.0.0/24` is the local nodeSubnet,
`174.16.1.0/24` is the CIDR of a BGP route.

```
-A POSTROUTING -s 10.128.0.0/24 -d 174.16.1.0/24 -j RETURN
-A POSTROUTING -s 10.128.0.0/24 -j MASQUERADE

> ```

Reviewer 4: I am not sure the route_import controller is the best place to do this. That controller is specific to OVN and already exceptional in that is IC only, both concerns which do not apply to LGW mode.

> ```

Reviewer 4: I think a new controller could be better. The only shared responsibility would be to subscribe to route updates and I am not sure that's enough to justify a mix.

    kind: FRRConfiguration
    metadata:
      name: cudn-1
      namespace: frr-k8s-system
      # A custom label
      label:
        network: default
    spec:
      bgp:
        routers:
        - asn: 64512
          neighbors:
          # the external BGP route reflector
          - address: 172.20.0.2
            asn: 64512
            disableMP: true
    ```

3. The cluster admin shall create a RouteAdvertisements CR to advertise the
   default network. Make sure the RouteAdvertisements CR selects the
   FRRConfiguration defined in the last step.

    ```yaml
    apiVersion: k8s.ovn.org/v1
    kind: RouteAdvertisements
    metadata:
      name: default
    spec:
      advertisements:
      - PodNetwork
      # Select the FFRConfiguration defined in step-2 with the custom label.
      frrConfigurationSelector:
        matchLabels:
          network: default
      networkSelectors:
      - networkSelectionType: DefaultNetwork
      # The empty nodeSelector selects all nodes. We don't support a network in a overlay and no-overlay hybrid mode.
      nodeSelector: {}
    ```

4. The cluster admin enables no-overlay mode by running ovn-kubernetes with the

>    configuration flag `--default-network-encapsulation=none` set.  

Author: No. If the flag is unset. OVN-K will operate in the geneve overlay mode by default.

>    configuration flag `--default-network-encapsulation=none` set.  

Author: Yes. The ovnkube-node pod shall exit with an error.

>    configuration flag `--default-network-encapsulation=none` set.  

Reviewer 6: do we need to consider OCP cluster here. Like what the steps need to prepare for day1 ? how to set the value to enable the no-overlay?  the external router is must before deploying the no-overlay cluster?  or create an internal Frrconfiguration with all nodeip as the neighbors if not external router . i.e

```
spec:
  bgp:
    routers:
    - asn: 64512
      vrf: blue
      neighbors:
      - address: 192.168.111.20
        asn: 64512
        toReceive:
          allowed:
            mode: all
      - address: 192.168.111.21
        asn: 64512
        toReceive:
          allowed:
            mode: all
```
  cno trigerr the frr-k8s and ovn-k with RA and frrconfiguration?  otherwise I guess the container pods cannot be communicate with different nodes. like api and auth operator may failed

>    configuration flag `--default-network-encapsulation=none` set.  

Author: @zhaozhanqi that can be addressed in the d/s enhancment. frr-k8s pods can be deployed in hostnetwork.

network for east-west traffic; this adds overhead and reduces throughput. By
leveraging OVN-Kubernetes support for BGP, we want to provide a way for users to
enable a no-overlay mode, which would disable Geneve encapsulation and use
direct routing between nodes for east-west traffic on selected networks.

Many environments, particularly on-premise deployments or those with dedicated
networking infrastructure, prefer to utilize the underlying physical network's
routing capabilities directly. This "no-overlay" approach can offer several
benefits:

* Improved Performance: Eliminates encapsulation/decapsulation overhead,
  potentially leading to lower latency and higher throughput for inter-pod
  communication.
* Simplified Troubleshooting: Traffic paths are more transparent as they align
  with the physical network's routing tables, simplifying debugging and network
  visibility.
* Leverage Existing Network Infrastructure: Integrates more seamlessly with
  existing BGP-capable network devices, allowing for direct routing to pod IPs.  
* Reduced Resource Consumption: Fewer CPU cycles spent on
  encapsulation/decapsulation.

## Goals

> * Support no-overlay mode for the default network.

Author: No-overlay will only be available for BGP-advertised networks. More details can be found in the subsequent chapters.

  First, in no-overlay mode, as there is no Geneve overhead, a network can have
  the same MTU size as the provider network. However, overlay networks require a
  smaller MTU. We cannot modify a pod interface MTU without restarting the pod.
  So OVN-Kubernetes must work with other external controllers to accommodate
  such a migration.
  Secondly, to facilitate a smooth, node-by-node migration, it's crucial to
  allow networks to operate in a hybrid no-overlay and overlay mode.  

* Support EgressIP and EgressService features in networks operating in
  no-overlay mode. Both features leverage OVN-Kubernetes's logical routing
  policy to steer pod egress traffic. They also rely on overlay tunnels to
  transport pod egress traffic to egress nodes. Implementing these features in
  no-overlay mode with BGP requires significant architectural changes and
  careful consideration.

* Support a network operating in an overlay and no-overlay hybrid mode.
  Currently, BGP only supports advertising entire pod networks. In the future,
  we may add support for selective node advertisement, which will enable hybrid
  modes. This will allow networks to operate in overlay and no-overlay modes

>   simultaneously. It is useful for scenarios such as overlay mode migration or

Reviewer 5: > > do we also have a use case for connecting half nodes using tunnels and others in underlay using bgp as a permanent mode?
> 
> It may be useful once we have EVPN.

not sure I see how its useful BUT I see this part was removed and we only mention migration, so resolving this.

  hybrid provider networks.

> ## Non-Goals

Author: It depends on whether route advertisements will be available for tenants or not.

* This enhancement does not aim to change the default behavior of
  OVN-Kubernetes, which will continue to use Geneve encapsulation for the
  default network and any user-defined networks unless no-overlay mode is
  explicitly enabled.

* This enhancement does not aim to change the existing CUDN/UDN isolation
  mechanism. CUDNs/UDNs will continue to be isolated from each other and the
  default network using OVN ACLs on every node logical switch.

* This enhancement does not aim to change the existing BGP configuration or
  behavior. The user must ensure that the BGP configuration is correctly set up
  to support no-overlay mode.

* This enhancement does not aim to change the existing CUDN lifecycle

>   management. The user must ensure that the CUDN CRs are correctly managed

Author: This means how users shall create/delete a CUDN. @ricky-rav

>   management. The user must ensure that the CUDN CRs are correctly managed

Reviewer 2: yeah, that's the idea.

  according to the existing lifecycle management practices.

* This enhancement does not aim to implement no-overlay mode with the
  centralized OVN architecture, as the BGP feature is only available in a
  cluster running in interconnect mode.

* This enhancement does not aim to implement no-overlay mode for the layer-2 or

>   localnet type of networks. The BGP feature doesn't support the localnet type

Author: It's about the CUDN configuration. Users shall not configure a CUDN in localnet with no-overlay enabled.

  of networks. The layer-2 type of networks is implemented by connecting pods to
  the same OVN distributed logical switch. Pods on different nodes are connected
  through a layer-2 segment using Geneve encapsulation. It's quite challenging
  to implement a layer-2 network over a layer-3 infrastructure without an
  overlay protocol.

> 

Author: We allow advertising CUDNs with overlapping pod subnets, as long as they are not advertised to the same VRF. This is the limitation of the BGP feature. If a user wants to turn on no-overlay mode for overlapping pod subnets, they shall configure a VRF-lite topology.

user-defined), allowing traffic to skip Geneve encapsulation (i.e., the overlay
network) and simply make use of the learned routes in the underlay or provider
network for inter-node communication.

While BGP is the proposed solution for exchanging east-west routing information
within a cluster, it is not the only possible option. Future implementations of
no-overlay mode may incorporate additional routing technologies and protocols
beyond BGP.

## User-Stories/Use-Cases

### Story 1: Enable no-overlay mode for the default network

As a cluster admin, I want to avoid all encapsulation for traffic in the default
network to integrate seamlessly with existing BGP-capable networks, achieve
maximum network performance, and simplify troubleshooting.

### Story 2: Enable no-overlay mode for a CUDN

As a cluster admin, I want to avoid all encapsulation for traffic in a CUDN to
integrate seamlessly with existing BGP-capable networks, achieve maximum network
performance, and simplify troubleshooting.

### Story 3: Enable no-overlay mode without an external BGP router

As a cluster admin, I want to avoid all encapsulation for traffic in the

> cluster, but I don't want to deploy an external BGP router for the cluster.

Author: This is just a deployment option for users who need no-overlay, but don't want to advertise the pod network to the external.
There is no 'default way' for this feature. Users shall choose the deployment mode based on their infrastructure and requirements.

> cluster, but I don't want to deploy an external BGP router for the cluster.

Reviewer 2: To integrate Peng's words:
```suggestion
cluster, but I don't want to advertise the pod network to the external and don't want to deploy an external BGP router for the cluster.
```

## Proposed Solution

The core idea is to leverage the existing BGP feature to advertise the IP subnet
allocated to each node (which contains the IPs of pods running on that node) to
an internal/external BGP route reflector. The BGP route reflector will then
populate these routes throughout the physical network, allowing other nodes to
directly route traffic to pod IPs without needing an overlay. Within a cluster,
different networks can operate in different overlay modes. Users can have
overlay and no-overlay networks simultaneously.

When no-overlay mode is enabled for a network:

* For north-south traffic, we will follow the existing implementation of the BGP
  feature; the traffic from pods will no longer be SNATed on egress. In the BGP
  feature, SNAT is disabled for all pod egress traffic that leaves the node,
  regardless of whether the destination route is BGP learned or not. For users
  who only want to have no-overlay mode but do not want to expose podIPs to the
  external, this behavior is not ideal. Instead of disabling SNAT for all the
  pod-to-external traffic, OVN-K shall adjust its SNAT behavior so that pod IPs

>   are exposed for traffic routed via BGP, but SNAT is still applied for

Author: No-overlay mode can only be set when a network is advertised. However, a network is advertised doesn't mean the network is advertised to all external networks. For example, a node can have a default route to a non-BGP router, and advertise the pod network to another BGP router. In this case, the outbond traffic to default route shall remain SNATed whereas the traffic to BGP router shall not be SNATed.

>   are exposed for traffic routed via BGP, but SNAT is still applied for

Reviewer 2: Uhm how do you tell whether the default route goes via a BGP router or not?

>   are exposed for traffic routed via BGP, but SNAT is still applied for

Author: If a route is learnt from a BGP router. I assume the nexthop is a BGP router.

  non-BGP-routed egress traffic.

* For east-west traffic, intra-node traffic (pod-to-pod, pod-to-clusterIP, and
  pod-to-host) remains unchanged, while cross-node traffic will follow the same

>   path as north-south traffic.

Reviewer 2: I think it's important to make this distinction extra clear in the document. 
So here we're saying that for clusters with no-overlay mode where the admin doesn't want to expose the pod subnets to the outside, node SNAT is still in place for egress traffic. Is this correct?

For the default network, we will add a new flag
`--default-network-encapsulation` to the `ovnkube` binary. This flag accepts
either `none` or `geneve`, with `geneve` being the default. Setting
`--default-network-encapsulation=none` will configure the default network to
operate in no-overlay mode. If the `--default-network-encapsulation=none` flag

> is set but the required RouteAdvertisements or FRRConfiguration CR is missing,

Author: This is to define the behavior when ovnkube starts `--default-network-encapsulation=none` but without `--enable-route-advertisements`.

> is set but the required RouteAdvertisements or FRRConfiguration CR is missing,

Reviewer 6: @pliurh 
> This is to define the behavior when ovnkube starts --default-network-encapsulation=none but without --enable-route-advertisement

is this mean if no-overlay is enabled for default network, but no need to --enable-route-advertisements ?  there will no routes in nodes host or GR router IIUC, is that expected?

> is set but the required RouteAdvertisements or FRRConfiguration CR is missing,

Reviewer 6: or as long as no-overlay enabled,  there will the related BGP routes generated in nodes host and GR router, then no need to apply RA cr?

encapsulation configuration cannot be changed after the
ClusterUserDefinedNetwork CR is created.

#### Example of a layer-3 CUDN with no-overlay mode enabled

A layer-3 CUDN with no-overlay mode enabled should look like this:

```yaml
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: my-cudn
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name
      operator: In
      values: ["red", "blue"]
  network:
    topology: Layer3
    layer3:
      role: Primary
      mtu: 1500
      subnets:
      - cidr: 10.10.0.0/16
        hostSubnet: 24
      encapsulation: "None"
```

The topology spec for each network type is shared between CUDN and UDN. Once we
add the new `encapsulation` field, it will be available for both CUDN and UDN.
To prevent a user from setting this field in a UDN CR accidentally, we will add
the following CEL validation rule to the UserDefinedNetwork CRD.

```yaml
    layer3:
      description: Layer3 is the Layer3 topology configuration.
      x-kubernetes-validations:
      - message: Setting layer3.encapsulation is not supported for UserDefinedNetworks
        rule: '!has(self.encapsulation)'

> ```

Author: I'll add a status condition for that.

> ```

Author: Added

between the transit switch and the ovn-cluster-router instances on each node.
That is the logical representation of the OVN network, and in practice, these
links are implemented through Geneve tunnels between every node in the cluster.
For a network to skip the encapsulation step and route pod traffic directly over
the underlay network, we need to change the logical network topology of this
network: its ovn-cluster-router won't connect to the transit switch anymore and
all east-west traffic will be naturally routed in the same path as north-south
traffic through the provider network.

The following diagram shows the OVN topology of a cluster with three networks:

* a CUDN (in green) that uses Geneve encapsulation (default behavior)  
* a default network (in gray) that is operating in no-overlay mode  
* a CUDN (in blue) that is operating in no-overlay mode

![No-overlay topology](../images/no_overlay_topology.jpg)

We need to update the ZoneInterconnectHandler for no-overlay mode. If a new
network is configured in no-overlay mode, ZoneInterconnectHandler shall not
create the transit switch for the network. For each remote zone, there is no

> need to add the static nodeSubnet routes into the ovn_cluster_router. The

Author: IIRC, @jcaamano suggested removing the transit switch for a clear topology.

traffic routed to those destination.

In the future, if any non-BGP routes in the gateway router have prefixes that
overlap with those learned via BGP, we can add these prefixes to another address
set and apply SNAT to traffic routed through them.

In the following example, `10.128.0.0/24` is the local nodeSubnet,
`174.16.1.0/24` is the prefix of a BGP route. `174.16.1.0/25` is the prefix of a
non-BGP route.

```text
# Address_Set contains addresses routed via BGP routers
_uuid               : 44e31651-db99-44f2-ad01-ff1d00bbb223
addresses           : ["174.16.1.0/24"]
external_ids        : {}
name                : skip_snat_dst

# Address_Set contains addresses routed via non-BGP routers
_uuid               : 55e31651-db99-44f2-ad01-ff1d00bbb223
addresses           : ["174.16.1.0/25"]
external_ids        : {}
name                : snat_dst

# SNAT rules in the Gateway Router, exclude traffic to BGP routers
_uuid               : cd9264de-f2c6-4528-8f53-c2ccb5b23c8d
allowed_ext_ips     : []
exempted_ext_ips    : []
external_ids        : {}
external_ip         : "172.18.0.2"
external_mac        : []
external_port_range : ""
gateway_port        : []

> logical_ip          : "10.244.1.7"

Author: Yeah, it shall be from 10.128.0.0/24

logical_port        : []

> match               : "ip4.dst!=$skip_snat_dst||ip4.dst==$snat_dst"

Author: `$skip_snat_dst` contains the prefixes of the imported BGP routes. I knew there would be a performance challenge. But I cannot think of a better approach.

> match               : "ip4.dst!=$skip_snat_dst||ip4.dst==$snat_dst"

Reviewer 2: If this is really a performance bottleneck, @jcaamano was suggesting on slack that we could derive the subnets for which we want to SNAT and apply it only to those:
https://redhat-internal.slack.com/archives/C090C97A90B/p1751622131107829?thread_ts=1751450578.459089&cid=C090C97A90B

I'm repeating this here because I've found it by chance and discussions are already scattered in multiple places :)

> match               : "ip4.dst!=$skip_snat_dst||ip4.dst==$snat_dst"

Author: That is the default behavior today for the overlay mode. So users don't need to configure NAT at the gateway router of the cluster if it's not a BGP router. It's also the default behavior of Calico BGP mode.

gateway router and the host routing table of each node. A new field
`encapsulation`, will be introduced to the NAD spec which indicates the
encapsulation of the network. If a network is operating in no-overlay mode, this
field will be set to `none`. This information can then be passed into the
NetInfo object and utilized by both the RouteAdvertisement controller and the
route import controller as described below.

An example OVN-Kubernetes NAD may look like:

```yaml
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: l3-network
  namespace: default
spec:
  config: |
    {
            "cniVersion": "0.3.1",
            "name": "l3-network",
            "type": "ovn-k8s-cni-overlay",
            "topology":"layer3",
            "mtu": 1500,
            "netAttachDefName": "default/l3-network",
            "role": "primary",
            "encapsulation": "none"
    }
```

##### Gateway Router

Currently, BGP routes to other node subnets are not imported into the gateway
router, which ensures node-to-pod traffic traverses the OVN logical topology.
However, we need the routes to be imported into the gateway router in no-overlay

> mode. We need to update the route import controller to check whether a network

Author: > if we fix the host-routing-table side via FRRConfig

I don't know what you mean by this.

is in no-overlay mode and handle the route importing accordingly.

##### Host Routing Tables

To [maintain isolation between UDNs](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5140),
node BGP speakers do not accept routes to NodeSubnets by default. However, to
route east-west traffic in no-overlay mode, these routes must be imported to the
host routing table. We need to update the node FRRConfiguration generation logic
to check whether a network is in no-overlay mode and generate the node
FRRConfiguration accordingly.

#### No SNAT for egress traffic

The OVN-Kubernetes BGP feature currently disables SNAT only for traffic destined
to external networks, not for inter-pod communication across nodes.

In no-overlay mode, the goal is to fully leverage BGP and eliminate the SNAT

> step on the gateway router, even for intra-cluster traffic. Within the same

Author: Pod network will always be advertised in the no-overlay mode.

(pod-to-other_node), to ensure nodePort services can be accessed across
networks. This egress traffic will remain SNATed to the node IP before leaving
the node. This is the current behavior when a network is advertised. We will not
change it for no-overlay mode.

#### Active traffic blocking between UDNs

The [existing UDN isolation
mechanism](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5186/commits/8f6e7d30ee5f4926a21e2de75488aad80344814b)
will still be in place. Pods in different UDNs will be isolated from each other
and the default network, utilizing the existing OVN ACLs on every node logical
switch: ACLs are enforced on outgoing traffic at each advertised UDN switch,
verifying whether both source and destination IPs belong to the same advertised
UDN. If the destination IP does not come from the same UDN subnet as the source
IP, traffic is dropped. No-overlay mode will not affect the isolation between
UDNs or between UDNs and the default network. So, even if multiple UDNs were
advertised to the same VRF, the cross-UDN traffic is still blocked. This is the
current behavior when a network is advertised. We will not change it for no

> overlay mode.

Author: The UDN isloation logic will not be affected by no-overlay mode with or without https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5276

> 

Reviewer 5: In the implementation section please flesh out what are the changes coming into:

1. cluster-manager
2. ovnkube-controller
3. ovnkube-node

components in detail - so either you can call them out in the above pieces or write them separately

### Workflow

#### Enable No-overlay Mode for the Default Network

1. The frr-k8s pod shall be deployed on each node.  
2. Before starting the ovnkube pods, the cluster admin shall create a base
   FRRConfiguration CR that is used for generating the per-node FRRConfiguration
   by OVN-K.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:
      name: default
      namespace: frr-k8s-system
      label:
        # A custom label
        network: default
    spec:
      bgp:
        routers:
        - asn: 64512
          neighbors:
          # the external BGP route reflector
          - address: 172.20.0.2
            asn: 64512
            disableMP: true
    ```

> 3. The cluster admin shall create a RouteAdvertisements CR to advertise the

Reviewer 5: hmm am I confusing something? I don't think creating RA CR is a requirement for no-overlay mode right? I think perhaps you mean to update the enhancement around this ? like we discussed in BGP sync meeting I think we want to create things internally for the user rather than expect them to have RA created etc.. which has a totally different API meaning.
I assumed RA today means:

1. advertise pod networks to outside the cluster elements

and this meaning of the API mustn't change IMHO

1. advertise pod networks to other nodes in the cluster -> this is the semantic of the new field you are adding which should require no additional end-user API changes

        clusterUserDefinedNetworkSelector:
          networkSelector:
            matchLabels:
              # Select the CUDN defined in step-2
              network: blue
      advertisements:
      - PodNetwork
    ```

5. OVN-Kubernetes will generate the following FRRConfiguration for each node.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:
      name: route-generated-blue
      namespace: frr-k8s-system
    spec:
      bgp:
        routers:
        - asn: 64512
          neighbors:
          - address: 182.18.0.5
            disableMP: true
            asn: 64512
            toAdvertise:
              allowed:
                prefixes:
                # the local node subnet
                - 10.10.1.0/24
            toReceive:
              # Allow to receive the routes to remote nodeSubnets.
              # This is no-overlay specific behavior.
              allowed:
                mode: filtered
                prefixes:
                - ge: 24
                  le: 24
                  prefix: 10.10.0.0/16
          prefixes:
          - 10.10.1.0/24
    ```

### Deployment Consideration

> #### BGP Topology

Reviewer 3: +1 this is a great section now, with awesome diagrams. Thanks!

The selected `FRRConfiguration` CR determines the deployment mode for the

> no-overlay network. There will be no new field added to the FRRConfiguration.

Reviewer 5: I think we are currently using RA API as the link to make changes - for me that's an extra user step.. shouldn't we do the configs ourself internally if the no-overlay-mode flag is set to none for that network?

and allows for more advanced routing policies (e.g., AS path prepending, MEDs,
and filtering based on AS paths).

In this topology, the FRR instance on each node establishes an eBGP session with
its upstream router. The upstream routers are then responsible for advertising
the node subnet routes to the rest of the network. It is also possible for
different groups of nodes (e.g. nodes in different racks) to belong to different
Autonomous Systems.

```mermaid
graph LR
    subgraph AS 65002 [AS 65002: Rack 1]
        direction LR
        NodeA(Node A)
        NodeB(Node B)
    end
    subgraph AS 65003 [AS 65003: Rack 2]
        direction LR
        NodeC(Node C)
    end
    subgraph AS 65001 [AS 65001: External Network]
        direction LR
        ToR1(ToR Switch 1)
        ToR2(ToR Switch 2)
    end
    NodeA <-- eBGP --> ToR1
    NodeB <-- eBGP --> ToR1
    NodeC <-- eBGP --> ToR2
    ToR1  <-- iBGP --> ToR2
```

##### Full-Mesh iBGP across the Cluster

The FRR instance on each node maintains a full mesh BGP peer relationship with
all other nodes across the cluster. In this mode, users can enable no-overlay

> mode without relying on external BGP routes.

Author: The workflow is the same as other deployments. The sole distinction lies in the BGP topology, as specified by the FRRConfiguration CR. I've given an example below.

        toReceive:
          allowed:
            mode: all
```

##### External Route Reflectors for Larger Clusters

In a large cluster, a full-mesh BGP setup leads to more CPU and memory
consumption on the nodes. Instead of every node peering with every other node,
nodes peer only with external BGP route reflectors. This significantly reduces
the number of BGP sessions each node needs to maintain, improving scalability.

```mermaid
graph TD
    subgraph AS 64512 [ AS 64512 ]
        direction LR
        subgraph "Route Reflector"
            RR1(Route Reflector 1)
        end
        subgraph "Kubernetes Nodes"
            NodeA(Node A)
            NodeB(Node B)
            NodeC(Node C)
        end
    end
    NodeA <-- iBGP --> RR1
    NodeB <-- iBGP --> RR1
    NodeC <-- iBGP --> RR1
```

##### Internal Route Reflectors

If an FRR node instance can act as a BGP route reflector, it enables the
creation of an internal route reflector-based BGP network. However, it is not
currently supported by frr-k8s, which we use for integrating OVN-K with FRR.
Therefore, this is not a possible deployment today.

### Feature Compatibility

#### Multiple External Gateways (MEG)

> The same as the route advertisement feature.

Reviewer 3: I think she means the -ds argument to kind:
```
-ds  | --disable-snat-multiple-gws            Disable SNAT for multiple gws. DEFAULT: Disabled.
```

If -ds is not used, then by default there is one SNAT for the entire pod subnet on the GR. If -ds is used, we create per pod SNATs. -ds is defaulted to true in OCP and required to use MEG.

> The same as the route advertisement feature.

Author: MEG cannot work with advertised networks, therefore MEG cannot work with networks operates in the no-overlay mode either.

# OKEP-5259: No-overlay Mode For Layer-3 Networks

* Issue: [\#5259](https://github.com/ovn-kubernetes/ovn-kubernetes/issues/5259)  
* Authors: Riccardo Ravaioli, Peng Liu

## Problem Statement

Currently, OVN-Kubernetes uses Geneve as its encapsulation method on the overlay
network for east-west traffic; this adds overhead and reduces throughput. By
leveraging OVN-Kubernetes support for BGP, we want to provide a way for users to
enable a no-overlay mode, which would disable Geneve encapsulation and use
direct routing between nodes for east-west traffic on selected networks.

Many environments, particularly on-premise deployments or those with dedicated
networking infrastructure, prefer to utilize the underlying physical network's
routing capabilities directly. This "no-overlay" approach can offer several
benefits:

* Improved Performance: Eliminates encapsulation/decapsulation overhead,
  potentially leading to lower latency and higher throughput for inter-pod
  communication.

> * Simplified Troubleshooting: Traffic paths are more transparent as they align

Author: Well, that's true. I'll remove this paragraph.

  with the physical network's routing tables, simplifying debugging and network
  visibility.
* Leverage Existing Network Infrastructure: Integrates more seamlessly with
  existing BGP-capable network devices, allowing for direct routing to pod IPs.  

> * Reduced Resource Consumption: Fewer CPU cycles spent on

Author: The statement here is "Fewer CPU cycles spent on encapsulation/decapsulation". That is 100% true, as there is no enca/decap anymore. I am not saying the overall CPU consumption will be less. The overall CPU consumption can be affected by many factors (route numbers, packet size, packet rate...). So it's hard to say whether the CPU cycles we save from enca/decap are more than the ovnk controller changes consumed.

  (CUDNs).
* A cluster can have networks operating in overlay and no-overlay modes
  simultaneously.
* Use the BGP feature to exchange routes to node subnets across the cluster.
* Allow direct communication without any overlay encapsulation for east-west
  traffic.
* Maintain compatibility with existing OVN-Kubernetes features where applicable.
* Compatible with both local gateway and shared gateway modes.

## Future Goals

* When OVN-Kubernetes supports BGP for UserDefinedNetwork CRs, extend no-overlay
  mode support for Primary layer-3 UDNs.

* Support toggling no-overlay mode on/off for an existing network. To support
  this function, there are two things we need to consider.
  First, in no-overlay mode, as there is no Geneve overhead, a network can have
  the same MTU size as the provider network. However, overlay networks require a
  smaller MTU. We cannot modify a pod interface MTU without restarting the pod.
  So OVN-Kubernetes must work with other external controllers to accommodate
  such a migration.
  Secondly, to facilitate a smooth, node-by-node migration, it's crucial to
  allow networks to operate in a hybrid no-overlay and overlay mode.  

> * Support EgressIP and EgressService features in networks operating in

Reviewer 5: @pliurh I think we dicussed this and agreed its too hard cause change needs to happen on the EgressIP and EgressService CRs which is complicated at this stage and out of scope and warrants its own epic? Was that reflected somewhere on the OKEP?

  no-overlay mode. Both features leverage OVN-Kubernetes's logical routing
  policy to steer pod egress traffic. They also rely on overlay tunnels to

>   transport pod egress traffic to egress nodes. Implementing these features in

Author: In the no-overlay mode, when the pod outbond traffic leaves its node, we have no control over how the next-hop routers forward the traffic, unless the nexthop is an ovn-k node. Therefore, we cannot ensure that the egress traffic can be forwarded to the egress IP node. That is the main difficulty of supporting EIP.

   * `encapsulation`: specifies the encapsulation method (default: `Geneve`)  
   * Supported values: `Geneve` (default) or `None` (no-overlay)

The `spec` field of a ClusterUserDefinedNetwork CR is immutable. Therefore, the
encapsulation configuration cannot be changed after the
ClusterUserDefinedNetwork CR is created.

#### Example of a layer-3 CUDN with no-overlay mode enabled

A layer-3 CUDN with no-overlay mode enabled should look like this:

```yaml
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: my-cudn
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name
      operator: In
      values: ["red", "blue"]
  network:
    topology: Layer3
    layer3:
      role: Primary
      mtu: 1500
      subnets:
      - cidr: 10.10.0.0/16
        hostSubnet: 24
      encapsulation: "None"
```

The topology spec for each network type is shared between CUDN and UDN. Once we
add the new `encapsulation` field, it will be available for both CUDN and UDN.
So, to prevent a user from setting this field in a UDN CR accidentally, a new
`status.conditions` type `NoOverlayEnabled` will be added to CUDN/UDN CRs. If a

> user tries to create a UDN CR with no-overlay mode enabled, cluster-manager

Reviewer 2: So did we dismiss the CEL in the end?

Author: CEL is part of the API. We want to support no-overlay UDN in the future, once BGP supports UDN. With this `status.conditions` approach, we don't need to change the API(CEL) when we support UDN. @jcaamano @tssurya WDYT?

Reviewer 4: For me CEL still looks like the better option

Reviewer 2: Yeah, so the CEL will be removed once we support no-overlay mode for UDN. Usually it's more confusing for a user to see their CR created and then realize that the cluster doesn't work as expected, rather than having the CR rejected from the start with a descriptive error message.

Reviewer 5: > CEL is part of the API. We want to support no-overlay UDN in the future, once BGP supports UDN. With this `status.conditions` approach, we don't need to change the API(CEL) when we support UDN. @jcaamano @tssurya WDYT?

yea having a CEL now to be removed later is fine

will still be in place. Pods in different UDNs will be isolated from each other
and the default network, utilizing the existing OVN ACLs on every node logical
switch: ACLs are enforced on outgoing traffic at each advertised UDN switch,
verifying whether both source and destination IPs belong to the same advertised
UDN. If the destination IP does not come from the same UDN subnet as the source
IP, traffic is dropped. No-overlay mode will not affect the isolation between
UDNs or between UDNs and the default network. So, even if multiple UDNs were
advertised to the same VRF, the cross-UDN traffic is still blocked. This is the
current behavior when a network is advertised. We will not change it for no
overlay mode.

### Workflow

#### Enable No-overlay Mode for the Default Network

1. The frr-k8s pod shall be deployed on each node.  
2. Before starting the ovnkube pods, the cluster admin shall create a base
   FRRConfiguration CR that is used for generating the per-node FRRConfiguration
   by OVN-K.

    ```yaml
    apiVersion: frrk8s.metallb.io/v1beta1
    kind: FRRConfiguration
    metadata:

>       name: cudn-1

Reviewer 2: nit: maybe let's call it `no-overlay-default` or something similar?

        toReceive:
          allowed:
            mode: all
      - address: 192.168.111.21
        asn: 64512
        disableMP: true
        toReceive:
          allowed:
            mode: all
      - address: 192.168.111.22
        asn: 64512
        disableMP: true
        toReceive:
          allowed:
            mode: all
```

##### External Route Reflectors for Larger Clusters

In a large cluster, a full-mesh BGP setup leads to more CPU and memory
consumption on the nodes. Instead of every node peering with every other node,
nodes peer only with external BGP route reflectors. This significantly reduces
the number of BGP sessions each node needs to maintain, improving scalability.

```mermaid
graph TD
    subgraph AS 64512 [ AS 64512 ]
        direction LR
        subgraph "Route Reflector"
            RR1(Route Reflector 1)
        end
        subgraph "Kubernetes Nodes"
            NodeA(Node A)
            NodeB(Node B)
            NodeC(Node C)
        end
    end
    NodeA <-- iBGP --> RR1
    NodeB <-- iBGP --> RR1
    NodeC <-- iBGP --> RR1
```

##### Internal Route Reflectors

While an FRR instance can act as a BGP route reflector, enabling an internal

> route reflector-based BGP network across the cluster. However, it is not

Reviewer 2: Is there a verb missing in the main sentence here?
